{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda2/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/setuptools/dist.py:331: UserWarning: Normalizing '1.0.1.1-dev' to '1.0.1.1.dev0'\n",
      "running develop\n",
      "running egg_info\n",
      "writing requirements to genepeeks.egg-info/requires.txt\n",
      "writing genepeeks.egg-info/PKG-INFO\n",
      "writing top-level names to genepeeks.egg-info/top_level.txt\n",
      "writing dependency_links to genepeeks.egg-info/dependency_links.txt\n",
      "reading manifest file 'genepeeks.egg-info/SOURCES.txt'\n",
      "writing manifest file 'genepeeks.egg-info/SOURCES.txt'\n",
      "running build_ext\n",
      "Creating /home/ec2-user/anaconda2/lib/python2.7/site-packages/genepeeks.egg-link (link to .)\n",
      "genepeeks 1.0.1.1.dev0 is already the active version in easy-install.pth\n",
      "\n",
      "Installed /home/ec2-user/git/genepeeks-science\n",
      "Processing dependencies for genepeeks==1.0.1.1.dev0\n",
      "Searching for apiclient==1.0.3\n",
      "Best match: apiclient 1.0.3\n",
      "Adding apiclient 1.0.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for config==0.3.9\n",
      "Best match: config 0.3.9\n",
      "Adding config 0.3.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for dropbox==7.1.1\n",
      "Best match: dropbox 7.1.1\n",
      "Adding dropbox 7.1.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for ggplot==0.6.8\n",
      "Best match: ggplot 0.6.8\n",
      "Adding ggplot 0.6.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for GitPython==2.1.0\n",
      "Best match: GitPython 2.1.0\n",
      "Adding GitPython 2.1.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for google-api-python-client==1.5.5\n",
      "Best match: google-api-python-client 1.5.5\n",
      "Adding google-api-python-client 1.5.5 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for oauth2client==1.5.2\n",
      "Best match: oauth2client 1.5.2\n",
      "Adding oauth2client 1.5.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for mock==2.0.0\n",
      "Best match: mock 2.0.0\n",
      "Adding mock 2.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for nose==1.3.7\n",
      "Best match: nose 1.3.7\n",
      "Adding nose 1.3.7 to easy-install.pth file\n",
      "Installing nosetests script to /home/ec2-user/anaconda2/bin\n",
      "Installing nosetests-2.7 script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for weightedstats==0.3\n",
      "Best match: weightedstats 0.3\n",
      "Adding weightedstats 0.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for seaborn==0.7.1\n",
      "Best match: seaborn 0.7.1\n",
      "Adding seaborn 0.7.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for matplotlib==1.5.2\n",
      "Best match: matplotlib 1.5.2\n",
      "Adding matplotlib 1.5.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for multiprocess==0.70.4\n",
      "Best match: multiprocess 0.70.4\n",
      "Adding multiprocess 0.70.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for contextlib2==0.5.4\n",
      "Best match: contextlib2 0.5.4\n",
      "Adding contextlib2 0.5.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pytz==2015.7\n",
      "Best match: pytz 2015.7\n",
      "Adding pytz 2015.7 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pymongo==3.3.0\n",
      "Best match: pymongo 3.3.0\n",
      "Adding pymongo 3.3.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for scipy==0.18.0\n",
      "Best match: scipy 0.18.0\n",
      "Adding scipy 0.18.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for numpy==1.11.1\n",
      "Best match: numpy 1.11.1\n",
      "Adding numpy 1.11.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for mando==0.4\n",
      "Best match: mando 0.4\n",
      "Adding mando 0.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for commandwrapper==0.7\n",
      "Best match: commandwrapper 0.7\n",
      "Adding commandwrapper 0.7 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for httplib2==0.9.2\n",
      "Best match: httplib2 0.9.2\n",
      "Adding httplib2 0.9.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for boto==2.42.0\n",
      "Best match: boto 2.42.0\n",
      "Adding boto 2.42.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for urllib3==1.20\n",
      "Best match: urllib3 1.20\n",
      "Adding urllib3 1.20 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for requests==2.12.4\n",
      "Best match: requests 2.12.4\n",
      "Adding requests 2.12.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for typing==3.5.3.0\n",
      "Best match: typing 3.5.3.0\n",
      "Adding typing 3.5.3.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for six==1.10.0\n",
      "Best match: six 1.10.0\n",
      "Adding six 1.10.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for brewer2mpl==1.4.1\n",
      "Best match: brewer2mpl 1.4.1\n",
      "Adding brewer2mpl 1.4.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for statsmodels==0.6.1\n",
      "Best match: statsmodels 0.6.1\n",
      "Adding statsmodels 0.6.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for patsy==0.4.1\n",
      "Best match: patsy 0.4.1\n",
      "Adding patsy 0.4.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pandas==0.18.1\n",
      "Best match: pandas 0.18.1\n",
      "Adding pandas 0.18.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for gitdb2==2.0.0\n",
      "Best match: gitdb2 2.0.0\n",
      "Adding gitdb2 2.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for uritemplate==3.0.0\n",
      "Best match: uritemplate 3.0.0\n",
      "Adding uritemplate 3.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyasn1==0.1.9\n",
      "Best match: pyasn1 0.1.9\n",
      "Adding pyasn1 0.1.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyasn1-modules==0.0.8\n",
      "Best match: pyasn1-modules 0.0.8\n",
      "Adding pyasn1-modules 0.0.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for rsa==3.4.2\n",
      "Best match: rsa 3.4.2\n",
      "Adding rsa 3.4.2 to easy-install.pth file\n",
      "Installing pyrsa-encrypt-bigfile script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-encrypt script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-verify script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-sign script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-priv2pub script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-decrypt script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-decrypt-bigfile script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-keygen script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pbr==1.10.0\n",
      "Best match: pbr 1.10.0\n",
      "Adding pbr 1.10.0 to easy-install.pth file\n",
      "Installing pbr script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for funcsigs==1.0.2\n",
      "Best match: funcsigs 1.0.2\n",
      "Adding funcsigs 1.0.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for python-dateutil==2.5.3\n",
      "Best match: python-dateutil 2.5.3\n",
      "Adding python-dateutil 2.5.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyparsing==2.1.4\n",
      "Best match: pyparsing 2.1.4\n",
      "Adding pyparsing 2.1.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for cycler==0.10.0\n",
      "Best match: cycler 0.10.0\n",
      "Adding cycler 0.10.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for dill==0.2.5\n",
      "Best match: dill 0.2.5\n",
      "Adding dill 0.2.5 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Sphinx==1.4.6\n",
      "Best match: Sphinx 1.4.6\n",
      "Processing Sphinx-1.4.6-py2.7.egg\n",
      "Sphinx 1.4.6 is already the active version in easy-install.pth\n",
      "Installing sphinx-apidoc script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-build script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-quickstart script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-autogen script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages/Sphinx-1.4.6-py2.7.egg\n",
      "Searching for smmap2==2.0.1\n",
      "Best match: smmap2 2.0.1\n",
      "Adding smmap2 2.0.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for imagesize==0.7.1\n",
      "Best match: imagesize 0.7.1\n",
      "Adding imagesize 0.7.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for alabaster==0.7.9\n",
      "Best match: alabaster 0.7.9\n",
      "Adding alabaster 0.7.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Babel==2.3.4\n",
      "Best match: Babel 2.3.4\n",
      "Adding Babel 2.3.4 to easy-install.pth file\n",
      "Installing pybabel script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for snowballstemmer==1.2.1\n",
      "Best match: snowballstemmer 1.2.1\n",
      "Adding snowballstemmer 1.2.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for docutils==0.12\n",
      "Best match: docutils 0.12\n",
      "Adding docutils 0.12 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Pygments==2.1.3\n",
      "Best match: Pygments 2.1.3\n",
      "Adding Pygments 2.1.3 to easy-install.pth file\n",
      "Installing pygmentize script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Jinja2==2.8\n",
      "Best match: Jinja2 2.8\n",
      "Adding Jinja2 2.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for MarkupSafe==0.23\n",
      "Best match: MarkupSafe 0.23\n",
      "Adding MarkupSafe 0.23 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Finished processing dependencies for genepeeks==1.0.1.1.dev0\n"
     ]
    }
   ],
   "source": [
    "! cd ../../genepeeks-science/ && python setup.py develop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from genepeeks.common import utilities as util\n",
    "import pysam, os\n",
    "import tabletext\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get DMD exons from gene collection in mongo using genepeeks science utilities function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished loading ensembl within the gene collection from the genepeeks-prod database with a single entries with a query of {'_id': 'DMD'} after 0.17 sec\n",
      "['Ex1', 'Ex2', 'Ex3', 'Ex4', 'Ex5', 'Ex6', 'Ex7', 'Ex8', 'Ex9', 'Ex10', 'Ex11', 'Ex12', 'Ex13', 'Ex14', 'Ex15', 'Ex16', 'Ex17', 'Ex18', 'Ex19', 'Ex20', 'Ex21', 'Ex22', 'Ex23', 'Ex24', 'Ex25', 'Ex26', 'Ex27', 'Ex28', 'Ex29', 'Ex30', 'Ex31', 'Ex32', 'Ex33', 'Ex34', 'Ex35', 'Ex36', 'Ex37', 'Ex38', 'Ex39', 'Ex40', 'Ex41', 'Ex42', 'Ex43', 'Ex44', 'Ex45', 'Ex46', 'Ex47', 'Ex48', 'Ex49', 'Ex50', 'Ex51', 'Ex52', 'Ex53', 'Ex54', 'Ex55', 'Ex56', 'Ex57', 'Ex58', 'Ex59', 'Ex60', 'Ex61', 'Ex62', 'Ex63', 'Ex64', 'Ex65|66', 'Ex67', 'Ex68', 'Ex69', 'Ex70', 'Ex71', 'Ex72', 'Ex73', 'Ex74', 'Ex75', 'Ex76', 'Ex77', 'Ex78', 'Ex79']\n"
     ]
    }
   ],
   "source": [
    "DMD_ensembl = util.Mongo.get_collection_data('gene', wanted_db='prod', query={'_id': 'DMD'}, find_one=True, single_field='ensembl')\n",
    "DMD_exons = util.get_nested_value(DMD_ensembl, ('is_primary', 'transcripts', 'is_primary', 'exons'))\n",
    "DMD_exons_merged = util.merge_intervals(DMD_exons, min_dist=200, include_index=True)\n",
    "exon_labels = ['Ex' + exon['index'] for exon in DMD_exons_merged]\n",
    "print exon_labels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get each sample's coverage across all DMD exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['subject' 'specimen' 'sample' ..., 'Ex77' 'Ex78' 'Ex79']\n",
      " ['MRMR-0011-266B' 'MRMR-0011-266B_1' 'MRMR-0011-266B_1-M1-1' ..., '70'\n",
      "  '13' '58']\n",
      " ['MRMR-0011-266B' 'MRMR-0011-266B_1' 'MRMR-0011-266B_1-M1-1' ..., '53'\n",
      "  '16' '55']\n",
      " ..., \n",
      " ['FRMR-00SD-679G' 'FRMR-00SD-679G_1' 'FRMR-00SD-679G_1-M1-1' ..., '58' '8'\n",
      "  '47']\n",
      " ['FRMR-00SD-679G' 'FRMR-00SD-679G_1' 'FRMR-00SD-679G_1-M1-1' ..., '73' '6'\n",
      "  '51']\n",
      " ['FRMR-00SD-679G' 'FRMR-00SD-679G_1' 'FRMR-00SD-679G_1-M1-1' ..., '66'\n",
      "  '14' '69']]\n"
     ]
    }
   ],
   "source": [
    "base_headers = ['subject', 'specimen', 'sample', 'gender', 'sequencer', 'flow_cell_id', 'lane']\n",
    "coverage_matrix = [base_headers + exon_labels]\n",
    "# print coverage_matrix\n",
    "# exit()\n",
    "for root, dirs, files in os.walk('/mnt/vep/subjects'):\n",
    "    for file_name in files:\n",
    "        if file_name.endswith('.bam'):# and file_name.startswith('F'):\n",
    "            # Initialize all exons to coverage of 0 for each subject\n",
    "#             subject_coverage = [0] * len(DMD_exons)\n",
    "#             subject = os.path.splitext(file_name)[0]\n",
    "            file_path = os.path.join(root, file_name)\n",
    "            subject_coverages = {}\n",
    "            bamfile = pysam.AlignmentFile(file_path, \"rb\")\n",
    "            for RG in bamfile.header['RG']:\n",
    "                subject, specimen_sample, flow_cell_lane = RG['ID'].split('_')\n",
    "                gender = subject[0]\n",
    "                specimen_num, sequencer, sample = specimen_sample.split('-')\n",
    "                specimen = '{}_{}'.format(subject, specimen_num)\n",
    "                sample = '{}_{}'.format(subject, specimen_sample)\n",
    "                flow_cell_id, lane = flow_cell_lane.rsplit('-', 1)\n",
    "                row = [subject, specimen, sample, gender, sequencer, flow_cell_id, lane]\n",
    "                subject_coverages[RG['ID']] = row + [0] * len(DMD_exons_merged)\n",
    "\n",
    "            for read in bamfile.fetch('X', start=31137345, end=33229636):\n",
    "                if not read.is_unmapped and read.mapping_quality == 60:\n",
    "                    # Find what exon each read falls in, and increase that exon's coverage by 1\n",
    "                    interval_info = util.in_interval(read.reference_start, DMD_exons_merged, get_interval=True)\n",
    "                    if not interval_info[0]:\n",
    "                        # If the start of the read is not in an exon, check the end of the read\n",
    "                        interval_info = util.in_interval(read.reference_end, DMD_exons_merged, get_interval=True)\n",
    "                    # else:\n",
    "                    #     # Make sure that the end of the read does not map to a different exon than the start\n",
    "                    #     end_interval_info = util.in_interval(read.reference_end, DMD_exons_merged, get_interval=True)\n",
    "                    #     if end_interval_info[0] and end_interval_info[1] != interval_info[1]:\n",
    "                    #         print read.reference_start, interval_info[1], read.reference_end, end_interval_info[1]\n",
    "                        \n",
    "                    if interval_info[0]:\n",
    "                        exon_num = interval_info[1]\n",
    "                        subject_coverages[read.get_tag('RG')][exon_num + len(base_headers)] += 1\n",
    "                        \n",
    "            coverage_matrix += subject_coverages.values()\n",
    "coverage_matrix = np.array(coverage_matrix)\n",
    "print coverage_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Print each exon's mean, std, min, and max coverage, and turn the stats into an array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "145 rows\n",
      "┌─────────┬────────┬────────┬─────┬─────┐\n",
      "│ Exon    │ Mean   │ Std    │ Min │ Max │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex1     │ 319.42 │ 112.96 │ 121 │ 649 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex2     │  12.02 │    6.3 │   1 │  33 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex3     │  76.68 │  28.02 │  27 │ 142 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex4     │ 129.99 │  51.16 │  44 │ 260 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex5     │  75.72 │  27.66 │  22 │ 143 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex6     │ 114.24 │  43.24 │  27 │ 228 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex7     │ 114.05 │  40.92 │  48 │ 212 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex8     │  82.72 │  30.86 │  30 │ 149 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex9     │  25.42 │  10.69 │   6 │  50 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex10    │ 107.52 │  43.47 │  25 │ 206 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex11    │  79.59 │  30.02 │  28 │ 148 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex12    │  67.56 │  25.87 │  18 │ 137 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex13    │  60.24 │  22.36 │  22 │ 117 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex14    │  46.14 │  18.13 │  12 │  83 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex15    │ 122.65 │   48.4 │  37 │ 235 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex16    │  50.25 │  18.99 │  15 │  87 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex17    │   65.6 │   24.6 │  20 │ 119 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex18    │  35.62 │  13.52 │  10 │  69 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex19    │   71.4 │  24.85 │  23 │ 127 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex20    │ 203.45 │  72.99 │  88 │ 368 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex21    │  87.22 │  33.04 │  29 │ 164 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex22    │ 111.53 │  42.24 │  43 │ 232 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex23    │  82.03 │  28.86 │  28 │ 153 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex24    │  66.33 │  23.35 │  23 │ 125 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex25    │ 192.56 │   68.3 │  73 │ 345 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex26    │  71.08 │  28.33 │  26 │ 136 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex27    │ 141.44 │  51.07 │  59 │ 288 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex28    │   62.1 │  22.22 │  18 │ 109 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex29    │ 108.41 │  39.77 │  33 │ 214 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex30    │  45.76 │  17.29 │  12 │  95 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex31    │  65.72 │  24.61 │  22 │ 128 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex32    │  70.08 │  26.55 │  28 │ 148 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex33    │  137.4 │  51.29 │  54 │ 252 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex34    │  58.13 │  20.63 │  14 │ 120 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex35    │  80.78 │   31.0 │  30 │ 171 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex36    │   78.8 │  28.02 │  26 │ 152 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex37    │  86.73 │  31.91 │  30 │ 182 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex38    │  57.87 │  22.81 │  15 │ 127 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex39    │  75.27 │  29.29 │  23 │ 140 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex40    │  85.76 │   30.9 │  29 │ 171 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex41    │  72.07 │  27.59 │  22 │ 144 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex42    │  53.43 │  19.74 │  16 │ 116 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex43    │ 109.48 │  42.88 │  31 │ 203 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex44    │  90.63 │  31.99 │  34 │ 179 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex45    │ 193.49 │  68.45 │  81 │ 364 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex46    │   63.4 │  23.01 │  22 │ 135 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex47    │ 160.29 │  60.68 │  68 │ 298 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex48    │  85.63 │  31.23 │  31 │ 150 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex49    │ 145.76 │  57.06 │  49 │ 276 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex50    │  60.72 │  22.94 │  20 │ 127 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex51    │  69.97 │  25.11 │  23 │ 152 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex52    │  51.13 │  17.71 │  11 │ 101 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex53    │ 137.24 │  51.23 │  42 │ 260 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex54    │  57.52 │  21.78 │  20 │ 118 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex55    │ 151.27 │  55.57 │  48 │ 262 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex56    │  48.19 │  18.13 │  14 │  88 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex57    │  50.15 │  18.26 │  17 │  89 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex58    │  57.02 │  22.12 │  15 │ 113 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex59    │  89.15 │  33.59 │  25 │ 175 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex60    │ 141.57 │  47.89 │  45 │ 254 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex61    │  56.16 │  20.56 │  19 │ 106 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex62    │ 111.24 │  37.55 │  41 │ 206 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex63    │  53.42 │  21.82 │  15 │ 117 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex64    │  89.43 │  33.77 │  31 │ 174 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex65|66 │  141.5 │  51.77 │  53 │ 257 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex67    │  82.52 │  30.52 │  32 │ 189 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex68    │ 126.92 │  46.82 │  46 │ 241 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex69    │  67.17 │  24.76 │  22 │ 128 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex70    │   84.2 │  33.91 │  25 │ 160 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex71    │  38.73 │  18.33 │   9 │  90 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex72    │  89.07 │  36.47 │  27 │ 172 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex73    │  54.43 │  19.86 │  17 │ 107 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex74    │ 112.21 │  39.68 │  42 │ 214 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex75    │   38.9 │  14.73 │  13 │  83 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex76    │  30.67 │  13.04 │   7 │  71 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex77    │  75.94 │   26.3 │  28 │ 133 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex78    │  13.58 │   5.56 │   1 │  30 │\n",
      "├─────────┼────────┼────────┼─────┼─────┤\n",
      "│ Ex79    │  75.22 │  28.82 │  18 │ 141 │\n",
      "└─────────┴────────┴────────┴─────┴─────┘\n"
     ]
    }
   ],
   "source": [
    "print len(coverage_matrix), 'rows'\n",
    "exon_cov_stats = [['Exon', 'Mean', 'Std', 'Min', 'Max']]\n",
    "exon_max_cov = []\n",
    "for i in range(len(DMD_exons_merged)):\n",
    "    col = i + len(base_headers)\n",
    "    exon_coverage = map(int, coverage_matrix[1:,col])\n",
    "    exon_cov_stats.append([coverage_matrix[0][col], round(np.mean(exon_coverage), 2), round(np.std(exon_coverage), 2), min(exon_coverage), max(exon_coverage)])\n",
    "# exon_stats_array = np.array(exon_cov_stats[1:,[1:5]])\n",
    "print tabletext.to_text(exon_cov_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Normalize each subject's coverage on each exon compared to the mean and std of that exon's coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.67197452  0.34883721  0.56981132 ...,  0.7337884   0.79245283\n",
      "   0.73668639]\n",
      " [ 0.65059145  0.41860465  0.55471698 ...,  0.6894198   0.49056604\n",
      "   0.73964497]\n",
      " [ 0.78207461  0.93023256  0.68301887 ...,  0.83959044  0.73584906\n",
      "   0.74556213]\n",
      " ..., \n",
      " [ 0.58826206  0.62790698  0.57735849 ...,  0.47098976  0.73584906\n",
      "   0.5443787 ]\n",
      " [ 0.51319381  0.37209302  0.35471698 ...,  0.49829352  0.58490566\n",
      "   0.53846154]\n",
      " [ 0.39126479  0.44186047  0.40754717 ...,  0.56313993  0.43396226\n",
      "   0.46153846]]\n",
      "[[-0.09158981 -1.14850299 -0.52688771 ..., -0.18404207  0.81959379\n",
      "   0.12059012]\n",
      " [-0.24263264 -0.78922156 -0.63833937 ..., -0.48125286 -1.09199522\n",
      "   0.14197135]\n",
      " [ 0.68612013  1.84550898  0.30899972 ...,  0.52469136  0.46117085\n",
      "   0.1847338 ]\n",
      " ..., \n",
      " [-0.68290645  0.28862275 -0.47116188 ..., -1.94444444  0.46117085\n",
      "  -1.26918965]\n",
      " [-1.21316322 -1.02874251 -2.11507384 ..., -1.7615455  -0.49462366\n",
      "  -1.31195211]\n",
      " [-2.07442877 -0.66946108 -1.72499303 ..., -1.32716049 -1.45041816\n",
      "  -1.86786402]]\n"
     ]
    }
   ],
   "source": [
    "coverage_matrix_vs_max_exon = coverage_matrix / exon_stats_array[:,-1]\n",
    "print coverage_matrix_vs_max_exon\n",
    "\n",
    "normalized_coverage_matrix = (coverage_matrix - exon_stats_array[:,1]) / exon_stats_array[:,2]\n",
    "print normalized_coverage_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get a matrix of the probability that any subject's reads will be for any particular exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['subject' 'specimen' 'sample' ..., 'Ex77' 'Ex78' 'Ex79']\n",
      " ['MRMR-0011-266B' 'MRMR-0011-266B_1' 'MRMR-0011-266B_1-M1-1' ...,\n",
      "  '0.0136213271064' '0.00252967503405' '0.0112862424596']\n",
      " ['MRMR-0011-266B' 'MRMR-0011-266B_1' 'MRMR-0011-266B_1-M1-1' ...,\n",
      "  '0.0104805220486' '0.0031639311845' '0.0108760134467']\n",
      " ..., \n",
      " ['FRMR-00SD-679G' 'FRMR-00SD-679G_1' 'FRMR-00SD-679G_1-M1-1' ...,\n",
      "  '0.0109186746988' '0.00150602409639' '0.00884789156627']\n",
      " ['FRMR-00SD-679G' 'FRMR-00SD-679G_1' 'FRMR-00SD-679G_1-M1-1' ...,\n",
      "  '0.0129570465034' '0.0010649627263' '0.00905218317359']\n",
      " ['FRMR-00SD-679G' 'FRMR-00SD-679G_1' 'FRMR-00SD-679G_1-M1-1' ...,\n",
      "  '0.012497633024' '0.00265101306571' '0.0130657072524']]\n"
     ]
    }
   ],
   "source": [
    "probability_in_an_exon_matrix = [list(coverage_matrix[0])]\n",
    "# print coverage_matrix.shape[1]\n",
    "# print coverage_matrix[1:, [(len(base_headers)+1):coverage_matrix.shape[1]-1]]\n",
    "# total_subject_reads = np.sum(map(int, coverage_matrix[1:,[len(base_headers)+1:-1]]), axis=1)\n",
    "for row in coverage_matrix[1:]:\n",
    "    total_subj_reads = sum(map(int, row[len(base_headers):]))\n",
    "    new_row = list(row[:len(base_headers)])\n",
    "    new_row += [float(item) / total_subj_reads for item in row[len(base_headers):]]\n",
    "    probability_in_an_exon_matrix.append(new_row)\n",
    "probability_in_an_exon_matrix = np.array(probability_in_an_exon_matrix)\n",
    "print probability_in_an_exon_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get summary stats for each subject across all exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "┌────────────────┬───────┬───────┬───────┬───────────┬──────────┬──────────┐\n",
      "│ Subject        │ Mean  │ Min   │ Max   │ Norm-Mean │ Norm-Min │ Norm-Max │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00SC-714V │ 0.677 │ 0.349 │ 0.986 │     -0.27 │   -1.378 │     1.44 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00CS-821F │ 0.711 │ 0.419 │ 0.917 │     -0.04 │   -1.471 │    1.261 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-0081-1987 │ 0.818 │ 0.534 │   1.0 │     0.683 │   -0.485 │    1.846 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-0085-579V │  0.66 │ 0.381 │ 0.856 │    -0.404 │   -1.817 │    0.765 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00DM-853C │ 0.606 │ 0.302 │ 0.782 │    -0.762 │   -2.083 │    0.175 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00NC-7632 │ 0.755 │ 0.488 │ 0.963 │     0.256 │   -0.872 │    1.554 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00GC-803L │ 0.963 │ 0.698 │   1.0 │     1.693 │    0.222 │    3.095 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00SA-876L │ 0.798 │ 0.511 │   1.0 │     0.554 │   -0.692 │    1.745 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00MS-704S │ 0.773 │ 0.442 │ 0.984 │     0.389 │   -1.092 │     1.42 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00KK-6426 │ 0.525 │  0.27 │  0.68 │    -1.317 │   -2.028 │   -0.224 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00TS-8464 │ 0.749 │ 0.534 │   1.0 │     0.211 │   -1.237 │    1.738 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00JM-865U │ 0.829 │ 0.582 │   1.0 │     0.773 │   -0.629 │    1.703 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00SB-851W │ 0.786 │ 0.464 │   1.0 │     0.468 │   -0.639 │    1.888 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00ID-7962 │ 0.769 │ 0.541 │ 0.995 │     0.341 │   -1.027 │    1.904 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00MM-839J │ 0.867 │ 0.465 │   1.0 │     1.018 │    -0.55 │    2.373 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00BB-772T │ 0.622 │ 0.321 │ 0.802 │    -0.643 │   -1.844 │    0.316 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00JD-7770 │ 0.661 │ 0.456 │  0.88 │    -0.393 │   -1.601 │    1.486 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-0070-444H │ 0.789 │ 0.572 │   1.0 │     0.487 │   -0.607 │    1.917 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00AW-8645 │ 0.872 │ 0.619 │   1.0 │     1.052 │    -0.31 │    2.474 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00ST-841F │ 0.568 │ 0.417 │ 0.736 │    -1.035 │   -1.944 │    0.461 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00AG-860T │  0.52 │ 0.298 │ 0.705 │    -1.364 │   -2.451 │   -0.371 │\n",
      "├────────────────┼───────┼───────┼───────┼───────────┼──────────┼──────────┤\n",
      "│ FRMR-00SD-679G │ 0.471 │ 0.238 │ 0.647 │    -1.698 │   -2.642 │   -0.669 │\n",
      "└────────────────┴───────┴───────┴───────┴───────────┴──────────┴──────────┘\n"
     ]
    }
   ],
   "source": [
    "subject_normalized_stats = [['Subject', 'Mean', 'Min', 'Max', 'Norm-Mean', 'Norm-Min', 'Norm-Max']]\n",
    "for i, subject in enumerate(subjects):\n",
    "    subject_stats = [subject, round(np.mean(coverage_matrix_vs_max_exon[i]), 3), round(min(coverage_matrix_vs_max_exon[i]), 3), round(max(coverage_matrix_vs_max_exon[i]), 3)]\n",
    "    subject_stats += [round(np.mean(normalized_coverage_matrix[i]), 3), round(min(normalized_coverage_matrix[i]), 3), round(max(normalized_coverage_matrix[i]), 3)]\n",
    "    subject_normalized_stats.append(subject_stats)\n",
    "print tabletext.to_text(subject_normalized_stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On branch master\r\n",
      "Your branch is up-to-date with 'origin/master'.\r\n",
      "Changes to be committed:\r\n",
      "  (use \"git reset HEAD <file>...\" to unstage)\r\n",
      "\r\n",
      "\t\u001b[32mmodified:   coverage_matrix.ipynb\u001b[m\r\n",
      "\r\n",
      "Changes not staged for commit:\r\n",
      "  (use \"git add <file>...\" to update what will be committed)\r\n",
      "  (use \"git checkout -- <file>...\" to discard changes in working directory)\r\n",
      "\r\n",
      "\t\u001b[31mmodified:   ../MAPQ_distribution.ipynb\u001b[m\r\n",
      "\t\u001b[31mmodified:   coverage_matrix.ipynb\u001b[m\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! git status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! git add coverage_matrix.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[master 2948311] Adding sample info to the coverage matrix\r\n",
      " 1 file changed, 942 insertions(+), 463 deletions(-)\r\n",
      " rewrite notebooks/coverage_matrix.ipynb (67%)\r\n"
     ]
    }
   ],
   "source": [
    "! git commit -m \"Adding sample info to the coverage matrix\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "warning: push.default is unset; its implicit value has changed in\n",
      "Git 2.0 from 'matching' to 'simple'. To squelch this message\n",
      "and maintain the traditional behavior, use:\n",
      "\n",
      "  git config --global push.default matching\n",
      "\n",
      "To squelch this message and adopt the new behavior now, use:\n",
      "\n",
      "  git config --global push.default simple\n",
      "\n",
      "When push.default is set to 'matching', git will push local branches\n",
      "to the remote branches that already exist with the same name.\n",
      "\n",
      "Since Git 2.0, Git defaults to the more conservative 'simple'\n",
      "behavior, which only pushes the current branch to the corresponding\n",
      "remote branch that 'git pull' uses to update the current branch.\n",
      "\n",
      "See 'git help config' and search for 'push.default' for further information.\n",
      "(the 'simple' mode was introduced in Git 1.7.11. Use the similar mode\n",
      "'current' instead of 'simple' if you sometimes use older versions of Git)\n",
      "\n",
      "Username for 'https://github.com': "
     ]
    }
   ],
   "source": [
    "! git push origin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
