{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda2/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/setuptools/dist.py:331: UserWarning: Normalizing '1.0.1.1-dev' to '1.0.1.1.dev0'\n",
      "running develop\n",
      "running egg_info\n",
      "writing requirements to genepeeks.egg-info/requires.txt\n",
      "writing genepeeks.egg-info/PKG-INFO\n",
      "writing top-level names to genepeeks.egg-info/top_level.txt\n",
      "writing dependency_links to genepeeks.egg-info/dependency_links.txt\n",
      "reading manifest file 'genepeeks.egg-info/SOURCES.txt'\n",
      "writing manifest file 'genepeeks.egg-info/SOURCES.txt'\n",
      "running build_ext\n",
      "Creating /home/ec2-user/anaconda2/lib/python2.7/site-packages/genepeeks.egg-link (link to .)\n",
      "genepeeks 1.0.1.1.dev0 is already the active version in easy-install.pth\n",
      "\n",
      "Installed /home/ec2-user/git/genepeeks-science\n",
      "Processing dependencies for genepeeks==1.0.1.1.dev0\n",
      "Searching for apiclient==1.0.3\n",
      "Best match: apiclient 1.0.3\n",
      "Adding apiclient 1.0.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for config==0.3.9\n",
      "Best match: config 0.3.9\n",
      "Adding config 0.3.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for dropbox==7.1.1\n",
      "Best match: dropbox 7.1.1\n",
      "Adding dropbox 7.1.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for ggplot==0.6.8\n",
      "Best match: ggplot 0.6.8\n",
      "Adding ggplot 0.6.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for GitPython==2.1.0\n",
      "Best match: GitPython 2.1.0\n",
      "Adding GitPython 2.1.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for google-api-python-client==1.5.5\n",
      "Best match: google-api-python-client 1.5.5\n",
      "Adding google-api-python-client 1.5.5 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for oauth2client==1.5.2\n",
      "Best match: oauth2client 1.5.2\n",
      "Adding oauth2client 1.5.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for mock==2.0.0\n",
      "Best match: mock 2.0.0\n",
      "Adding mock 2.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for nose==1.3.7\n",
      "Best match: nose 1.3.7\n",
      "Adding nose 1.3.7 to easy-install.pth file\n",
      "Installing nosetests script to /home/ec2-user/anaconda2/bin\n",
      "Installing nosetests-2.7 script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for weightedstats==0.3\n",
      "Best match: weightedstats 0.3\n",
      "Adding weightedstats 0.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for seaborn==0.7.1\n",
      "Best match: seaborn 0.7.1\n",
      "Adding seaborn 0.7.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for matplotlib==1.5.2\n",
      "Best match: matplotlib 1.5.2\n",
      "Adding matplotlib 1.5.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for multiprocess==0.70.4\n",
      "Best match: multiprocess 0.70.4\n",
      "Adding multiprocess 0.70.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for contextlib2==0.5.4\n",
      "Best match: contextlib2 0.5.4\n",
      "Adding contextlib2 0.5.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pytz==2015.7\n",
      "Best match: pytz 2015.7\n",
      "Adding pytz 2015.7 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pymongo==3.3.0\n",
      "Best match: pymongo 3.3.0\n",
      "Adding pymongo 3.3.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for scipy==0.18.0\n",
      "Best match: scipy 0.18.0\n",
      "Adding scipy 0.18.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for numpy==1.11.1\n",
      "Best match: numpy 1.11.1\n",
      "Adding numpy 1.11.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for mando==0.4\n",
      "Best match: mando 0.4\n",
      "Adding mando 0.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for commandwrapper==0.7\n",
      "Best match: commandwrapper 0.7\n",
      "Adding commandwrapper 0.7 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for httplib2==0.9.2\n",
      "Best match: httplib2 0.9.2\n",
      "Adding httplib2 0.9.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for boto==2.42.0\n",
      "Best match: boto 2.42.0\n",
      "Adding boto 2.42.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for urllib3==1.20\n",
      "Best match: urllib3 1.20\n",
      "Adding urllib3 1.20 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for requests==2.12.4\n",
      "Best match: requests 2.12.4\n",
      "Adding requests 2.12.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for typing==3.5.3.0\n",
      "Best match: typing 3.5.3.0\n",
      "Adding typing 3.5.3.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for six==1.10.0\n",
      "Best match: six 1.10.0\n",
      "Adding six 1.10.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for brewer2mpl==1.4.1\n",
      "Best match: brewer2mpl 1.4.1\n",
      "Adding brewer2mpl 1.4.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for statsmodels==0.6.1\n",
      "Best match: statsmodels 0.6.1\n",
      "Adding statsmodels 0.6.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for patsy==0.4.1\n",
      "Best match: patsy 0.4.1\n",
      "Adding patsy 0.4.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pandas==0.18.1\n",
      "Best match: pandas 0.18.1\n",
      "Adding pandas 0.18.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for gitdb2==2.0.0\n",
      "Best match: gitdb2 2.0.0\n",
      "Adding gitdb2 2.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for uritemplate==3.0.0\n",
      "Best match: uritemplate 3.0.0\n",
      "Adding uritemplate 3.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyasn1==0.1.9\n",
      "Best match: pyasn1 0.1.9\n",
      "Adding pyasn1 0.1.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyasn1-modules==0.0.8\n",
      "Best match: pyasn1-modules 0.0.8\n",
      "Adding pyasn1-modules 0.0.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for rsa==3.4.2\n",
      "Best match: rsa 3.4.2\n",
      "Adding rsa 3.4.2 to easy-install.pth file\n",
      "Installing pyrsa-encrypt-bigfile script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-encrypt script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-verify script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-sign script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-priv2pub script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-decrypt script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-decrypt-bigfile script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-keygen script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pbr==1.10.0\n",
      "Best match: pbr 1.10.0\n",
      "Adding pbr 1.10.0 to easy-install.pth file\n",
      "Installing pbr script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for funcsigs==1.0.2\n",
      "Best match: funcsigs 1.0.2\n",
      "Adding funcsigs 1.0.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for python-dateutil==2.5.3\n",
      "Best match: python-dateutil 2.5.3\n",
      "Adding python-dateutil 2.5.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyparsing==2.1.4\n",
      "Best match: pyparsing 2.1.4\n",
      "Adding pyparsing 2.1.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for cycler==0.10.0\n",
      "Best match: cycler 0.10.0\n",
      "Adding cycler 0.10.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for dill==0.2.5\n",
      "Best match: dill 0.2.5\n",
      "Adding dill 0.2.5 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Sphinx==1.4.6\n",
      "Best match: Sphinx 1.4.6\n",
      "Processing Sphinx-1.4.6-py2.7.egg\n",
      "Sphinx 1.4.6 is already the active version in easy-install.pth\n",
      "Installing sphinx-apidoc script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-build script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-quickstart script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-autogen script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages/Sphinx-1.4.6-py2.7.egg\n",
      "Searching for smmap2==2.0.1\n",
      "Best match: smmap2 2.0.1\n",
      "Adding smmap2 2.0.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for imagesize==0.7.1\n",
      "Best match: imagesize 0.7.1\n",
      "Adding imagesize 0.7.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for alabaster==0.7.9\n",
      "Best match: alabaster 0.7.9\n",
      "Adding alabaster 0.7.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Babel==2.3.4\n",
      "Best match: Babel 2.3.4\n",
      "Adding Babel 2.3.4 to easy-install.pth file\n",
      "Installing pybabel script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for snowballstemmer==1.2.1\n",
      "Best match: snowballstemmer 1.2.1\n",
      "Adding snowballstemmer 1.2.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for docutils==0.12\n",
      "Best match: docutils 0.12\n",
      "Adding docutils 0.12 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Pygments==2.1.3\n",
      "Best match: Pygments 2.1.3\n",
      "Adding Pygments 2.1.3 to easy-install.pth file\n",
      "Installing pygmentize script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Jinja2==2.8\n",
      "Best match: Jinja2 2.8\n",
      "Adding Jinja2 2.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for MarkupSafe==0.23\n",
      "Best match: MarkupSafe 0.23\n",
      "Adding MarkupSafe 0.23 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Finished processing dependencies for genepeeks==1.0.1.1.dev0\n"
     ]
    }
   ],
   "source": [
    "! cd ../../genepeeks-science/ && python setup.py develop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from genepeeks.common import utilities as util\n",
    "import pysam, os\n",
    "import tabletext\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get DMD exons from gene collection in mongo using genepeeks science utilities function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished loading ensembl within the gene collection from the genepeeks-prod database with a single entries with a query of {'_id': 'DMD'} after 0.12 sec\n",
      "['Ex1', 'Ex2', 'Ex3', 'Ex4', 'Ex5', 'Ex6', 'Ex7', 'Ex8', 'Ex9', 'Ex10', 'Ex11', 'Ex12', 'Ex13', 'Ex14', 'Ex15', 'Ex16', 'Ex17', 'Ex18', 'Ex19', 'Ex20', 'Ex21', 'Ex22', 'Ex23', 'Ex24', 'Ex25', 'Ex26', 'Ex27', 'Ex28', 'Ex29', 'Ex30', 'Ex31', 'Ex32', 'Ex33', 'Ex34', 'Ex35', 'Ex36', 'Ex37', 'Ex38', 'Ex39', 'Ex40', 'Ex41', 'Ex42', 'Ex43', 'Ex44', 'Ex45', 'Ex46', 'Ex47', 'Ex48', 'Ex49', 'Ex50', 'Ex51', 'Ex52', 'Ex53', 'Ex54', 'Ex55', 'Ex56', 'Ex57', 'Ex58', 'Ex59', 'Ex60', 'Ex61', 'Ex62', 'Ex63', 'Ex64', 'Ex65|66', 'Ex67', 'Ex68', 'Ex69', 'Ex70', 'Ex71', 'Ex72', 'Ex73', 'Ex74', 'Ex75', 'Ex76', 'Ex77', 'Ex78', 'Ex79']\n"
     ]
    }
   ],
   "source": [
    "DMD_ensembl = util.Mongo.get_collection_data('gene', wanted_db='prod', query={'_id': 'DMD'}, find_one=True, single_field='ensembl')\n",
    "DMD_exons = util.get_nested_value(DMD_ensembl, ('is_primary', 'transcripts', 'is_primary', 'exons'))\n",
    "DMD_exons_merged = util.merge_intervals(DMD_exons, min_dist=200, include_index=True)\n",
    "exon_labels = ['Ex' + exon['index'] for exon in DMD_exons_merged]\n",
    "print exon_labels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Count the number of males and females"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "173\n",
      "256\n"
     ]
    }
   ],
   "source": [
    "F_count = M_count = 0\n",
    "for root, dirs, files in os.walk('/mnt/vep/subjects'):\n",
    "    for file_name in files:\n",
    "        if file_name.endswith('.bam'):\n",
    "            if file_name.startswith('F'):\n",
    "                F_count += 1\n",
    "            elif file_name.startswith('M'):\n",
    "                M_count += 1\n",
    "print F_count\n",
    "print M_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get each sample's coverage across all DMD exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FK2R-0000-151I.bam is missing an index file\n",
      "MCLR-NA12-8911.bam is missing an index file\n",
      "[['id' 'subject' 'specimen' ..., 'Ex77' 'Ex78' 'Ex79']\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L001' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 71 12 81]\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L002' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 71 14 67]\n",
      " ..., \n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L003' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 131 19 144]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L002' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 128 22 117]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L004' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 139 25 126]]\n"
     ]
    }
   ],
   "source": [
    "base_headers = ['id', 'subject', 'specimen', 'sample', 'gender', 'sequencer', 'flow_cell_id', 'lane', 'bwa_version']\n",
    "coverage_matrix = [base_headers + exon_labels]\n",
    "for root, dirs, files in os.walk('/mnt/vep/subjects'):\n",
    "    for file_name in files:\n",
    "        if file_name.endswith('.bam'):\n",
    "            if '{}.bai'.format(file_name) not in files:\n",
    "                print '{} is missing an index file'.format(file_name)\n",
    "                continue\n",
    "            # Initialize all exons to coverage of 0 for each subject\n",
    "#             subject_coverage = [0] * len(DMD_exons)\n",
    "#             subject = os.path.splitext(file_name)[0]\n",
    "            file_path = os.path.join(root, file_name)\n",
    "            subject_coverages = {}\n",
    "            bamfile = pysam.AlignmentFile(file_path, \"rb\")\n",
    "            bwa_version = next(PG['VN'] for PG in bamfile.header['PG'] if PG.get('ID') == 'bwa')\n",
    "            for RG in bamfile.header['RG']:\n",
    "                try:\n",
    "                    # normal RG['ID'] format: FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L004\n",
    "                    subject, specimen_sample, flow_cell_lane = RG['ID'].split('_')\n",
    "                except:\n",
    "                    # older RG['ID'] format: FPWB-0000-429L_1-P1-1\n",
    "                    subject, specimen_sample = RG['ID'].split('_')\n",
    "                    flow_cell_id = lane = None\n",
    "                else:\n",
    "                    flow_cell_id, lane = flow_cell_lane.rsplit('-', 1)\n",
    "\n",
    "                gender = subject[0]\n",
    "                if specimen_sample.startswith(('ACGT', 'Omega')):\n",
    "                    lab, specimen_num, sequencer, sample = specimen_sample.split('.')\n",
    "                    specimen_num = '{}_{}'.format(lab, specimen_num)\n",
    "                else:\n",
    "                    specimen_num, sequencer, sample = specimen_sample.split('-')\n",
    "                specimen = '{}_{}'.format(subject, specimen_num)\n",
    "                sample = '{}_{}'.format(subject, specimen_sample)\n",
    "                row = [RG['ID'], subject, specimen, sample, gender, sequencer, flow_cell_id, lane, bwa_version]\n",
    "                subject_coverages[RG['ID']] = row + [0] * len(DMD_exons_merged)\n",
    "\n",
    "            for read in bamfile.fetch('X', start=31137345, end=33229636):\n",
    "                if not read.is_unmapped and read.mapping_quality == 60:\n",
    "                    # Find what exon each read falls in, and increase that exon's coverage by 1\n",
    "                    interval_info = util.in_interval(read.reference_start, DMD_exons_merged, get_interval=True)\n",
    "                    if not interval_info[0]:\n",
    "                        # If the start of the read is not in an exon, check the end of the read\n",
    "                        interval_info = util.in_interval(read.reference_end, DMD_exons_merged, get_interval=True)\n",
    "#                     else:\n",
    "#                         # Make sure that the end of the read does not map to a different exon than the start\n",
    "#                         end_interval_info = util.in_interval(read.reference_end, DMD_exons_merged, get_interval=True)\n",
    "#                         if end_interval_info[0] and end_interval_info[1] != interval_info[1]:\n",
    "#                             print read.reference_start, interval_info[1], read.reference_end, end_interval_info[1]\n",
    "                        \n",
    "                    if interval_info[0]:\n",
    "                        exon_num = interval_info[1]\n",
    "                        subject_coverages[read.get_tag('RG')][exon_num + len(base_headers)] += 1\n",
    "                        \n",
    "            coverage_matrix += subject_coverages.values()\n",
    "coverage_matrix = np.array(coverage_matrix)\n",
    "print coverage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.DataFrame(coverage_matrix)\n",
    "df.to_csv(\"../exon_data/coverage_matrix.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Print each exon's mean, std, min, and max coverage, and turn the stats into an array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1210 rows\n",
      "┌──────┬────────┬────────┬─────┬───────┬────────┬──────┐\n",
      "│ Exon │ Mean   │ Std    │ Min │ 1%    │ 99%    │ Max  │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  245 │  431.3 │ 340.84 │   0 │ 39.18 │ 2123.1 │ 2647 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   13 │  14.57 │   9.02 │   0 │   0.0 │  38.46 │   42 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   54 │ 102.78 │  54.76 │   0 │ 24.36 │  291.3 │  476 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  115 │ 174.62 │ 118.71 │   0 │ 41.18 │ 725.58 │  982 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   53 │ 111.47 │  67.91 │   0 │ 29.18 │  414.2 │  552 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  108 │ 150.52 │  85.78 │   0 │ 25.18 │ 490.82 │  654 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   71 │ 127.28 │  85.54 │   0 │  4.18 │ 331.84 │  461 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   39 │  94.49 │  62.74 │   0 │  1.36 │  265.6 │  370 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   16 │  34.29 │   19.8 │   0 │   3.0 │  91.28 │  110 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   85 │ 130.81 │  82.22 │   0 │  16.9 │ 409.48 │  535 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  110 │ 113.81 │  42.75 │   0 │  32.0 │ 229.46 │  295 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   75 │ 103.14 │   47.0 │   0 │  33.0 │ 294.44 │  385 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   97 │ 106.02 │  49.07 │   0 │ 32.18 │ 273.28 │  407 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   52 │  55.93 │  29.79 │   0 │   6.0 │ 119.82 │  144 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  118 │  154.9 │  72.52 │   0 │ 31.18 │ 354.94 │  443 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   45 │  69.09 │  41.87 │   0 │ 14.18 │ 251.02 │  323 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   68 │  83.82 │  47.79 │   0 │  7.18 │  226.2 │  283 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   47 │  50.36 │  23.73 │   0 │  10.0 │ 111.64 │  146 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   54 │  87.25 │   58.1 │   0 │  4.18 │ 274.28 │  344 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  146 │ 249.89 │ 125.75 │   0 │ 60.08 │ 617.58 │  818 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   89 │ 122.33 │  67.48 │   0 │ 23.18 │  391.1 │  544 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   83 │ 158.96 │   94.3 │   0 │  44.0 │ 613.58 │  685 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  100 │ 127.79 │  58.75 │   0 │  33.9 │ 314.82 │  434 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   42 │  93.06 │  73.73 │   0 │  8.54 │ 470.36 │  545 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  129 │ 228.15 │ 159.76 │   0 │ 20.36 │ 773.66 │ 1026 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   71 │ 100.24 │  77.29 │   0 │ 16.36 │ 477.96 │  608 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  122 │ 182.53 │ 144.45 │   0 │ 24.36 │  884.0 │ 1099 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   62 │  76.35 │  42.81 │   0 │  8.54 │  184.0 │  260 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  137 │ 175.31 │   83.0 │   0 │ 61.36 │ 476.88 │  674 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   33 │   60.9 │  37.12 │   0 │  7.54 │ 202.12 │  304 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   54 │  74.04 │  42.87 │   0 │  4.18 │ 168.82 │  201 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   71 │ 118.52 │ 105.04 │   0 │ 28.36 │  713.6 │  847 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  146 │ 182.02 │  92.04 │   0 │ 42.18 │ 503.94 │  666 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   64 │   94.5 │  60.78 │   0 │ 25.18 │  391.5 │  473 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   92 │ 123.08 │  57.77 │   0 │ 35.72 │ 321.64 │  406 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  106 │ 119.52 │  55.45 │   0 │ 35.54 │ 335.48 │  452 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   81 │ 133.97 │  79.31 │   0 │ 33.72 │ 486.22 │  651 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   61 │   79.5 │  44.68 │   0 │  11.9 │ 257.74 │  339 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  105 │ 120.43 │  52.68 │   0 │ 43.54 │ 317.46 │  351 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  124 │ 124.18 │  52.14 │   0 │ 26.36 │ 278.94 │  377 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  100 │ 112.33 │  56.73 │   0 │ 33.72 │ 312.12 │  452 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   64 │  87.46 │  57.35 │   0 │ 20.18 │ 373.76 │  443 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  140 │ 164.08 │  70.09 │   0 │ 45.08 │ 402.82 │  628 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   52 │ 116.15 │  90.44 │   0 │ 10.18 │ 489.26 │  592 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  105 │ 220.87 │ 150.51 │   0 │ 16.72 │ 643.24 │  826 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   60 │   99.3 │  64.67 │   0 │ 22.72 │ 387.64 │  533 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   97 │ 189.28 │ 133.87 │   0 │  18.0 │ 700.88 │  873 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  134 │ 130.55 │   70.4 │   0 │ 38.18 │ 462.58 │  555 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   94 │ 156.48 │ 102.86 │   0 │  8.54 │ 387.64 │  469 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   80 │  87.27 │  46.27 │   0 │ 19.18 │  263.6 │  347 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   88 │ 114.76 │  58.29 │   0 │ 33.72 │ 321.76 │  465 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   81 │  80.05 │  43.69 │   0 │ 23.72 │  273.1 │  345 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   83 │ 176.57 │  99.99 │   0 │ 38.18 │ 604.34 │  708 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   72 │  84.19 │  46.71 │   0 │ 15.72 │ 253.28 │  384 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  111 │ 200.14 │ 105.96 │   0 │ 50.72 │ 623.44 │  761 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   94 │  75.07 │  33.39 │   0 │  20.0 │ 214.82 │  247 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   53 │  81.02 │  52.15 │   0 │ 18.54 │ 321.58 │  478 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   86 │ 100.83 │   50.8 │   0 │ 29.54 │ 269.82 │  438 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  126 │ 135.45 │   56.1 │   0 │  41.0 │ 342.32 │  426 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  109 │ 190.52 │ 147.44 │   0 │ 21.72 │  877.4 │ 1095 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   33 │  76.78 │  51.71 │   0 │ 14.18 │  314.1 │  387 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   92 │ 143.77 │  87.71 │   0 │ 16.26 │ 476.84 │  586 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   68 │  84.84 │  62.81 │   0 │  21.0 │  391.4 │  496 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   74 │ 110.36 │  61.59 │   0 │ 18.18 │ 244.82 │  359 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   98 │ 199.77 │ 137.49 │   0 │ 48.36 │ 863.66 │  996 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   98 │ 123.38 │   64.2 │   0 │ 28.26 │ 375.68 │  476 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  100 │ 142.44 │  85.24 │   0 │  11.9 │  339.1 │  359 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   74 │  109.5 │  73.51 │   0 │ 28.18 │ 486.48 │  566 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  102 │  135.8 │  61.03 │   0 │ 46.18 │ 330.28 │  527 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   42 │  51.97 │  23.03 │   0 │ 10.18 │ 125.38 │  160 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   76 │ 111.83 │  62.78 │   0 │ 12.72 │ 347.66 │  408 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   55 │  66.09 │  35.01 │   0 │   9.0 │ 153.28 │  222 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│  139 │ 171.62 │  90.82 │   0 │ 56.18 │ 575.92 │  720 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   44 │  55.34 │  42.67 │   0 │   6.0 │ 255.92 │  298 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   30 │  45.74 │  20.54 │   0 │  10.0 │  115.0 │  147 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   71 │ 106.66 │  65.21 │   0 │ 20.18 │ 417.68 │  482 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   12 │  20.89 │   19.6 │   0 │  0.18 │  115.2 │  152 │\n",
      "├──────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│   81 │  99.68 │  58.32 │   0 │  18.0 │  310.6 │  399 │\n",
      "└──────┴────────┴────────┴─────┴───────┴────────┴──────┘\n"
     ]
    }
   ],
   "source": [
    "print len(coverage_matrix), 'rows'\n",
    "exon_cov_stats = [['Exon', 'Mean', 'Std', 'Min', '1%', '99%', 'Max']]\n",
    "exon_max_cov = []\n",
    "female_coverage_matrix = np.array([list(coverage_matrix[0])] + [list(item) for item in coverage_matrix if item[0].startswith('F')])\n",
    "for i in range(len(DMD_exons_merged)):\n",
    "    col = i + len(base_headers)\n",
    "    exon_coverage = map(int, female_coverage_matrix[1:,col])\n",
    "    exon_stats = [\n",
    "        coverage_matrix[0][col],\n",
    "        round(np.mean(exon_coverage), 2),\n",
    "        round(np.std(exon_coverage), 2),\n",
    "        min(exon_coverage),\n",
    "        np.percentile(exon_coverage, 1),\n",
    "        np.percentile(exon_coverage, 99),\n",
    "        max(exon_coverage)\n",
    "    ]\n",
    "    exon_cov_stats.append(exon_stats)\n",
    "# exon_stats_array = np.array(exon_cov_stats[1:,[1:5]])\n",
    "print tabletext.to_text(exon_cov_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Normalize each subject's coverage on each exon compared to the mean and std of that exon's coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.67197452  0.34883721  0.56981132 ...,  0.7337884   0.79245283\n",
      "   0.73668639]\n",
      " [ 0.65059145  0.41860465  0.55471698 ...,  0.6894198   0.49056604\n",
      "   0.73964497]\n",
      " [ 0.78207461  0.93023256  0.68301887 ...,  0.83959044  0.73584906\n",
      "   0.74556213]\n",
      " ..., \n",
      " [ 0.58826206  0.62790698  0.57735849 ...,  0.47098976  0.73584906\n",
      "   0.5443787 ]\n",
      " [ 0.51319381  0.37209302  0.35471698 ...,  0.49829352  0.58490566\n",
      "   0.53846154]\n",
      " [ 0.39126479  0.44186047  0.40754717 ...,  0.56313993  0.43396226\n",
      "   0.46153846]]\n",
      "[[-0.09158981 -1.14850299 -0.52688771 ..., -0.18404207  0.81959379\n",
      "   0.12059012]\n",
      " [-0.24263264 -0.78922156 -0.63833937 ..., -0.48125286 -1.09199522\n",
      "   0.14197135]\n",
      " [ 0.68612013  1.84550898  0.30899972 ...,  0.52469136  0.46117085\n",
      "   0.1847338 ]\n",
      " ..., \n",
      " [-0.68290645  0.28862275 -0.47116188 ..., -1.94444444  0.46117085\n",
      "  -1.26918965]\n",
      " [-1.21316322 -1.02874251 -2.11507384 ..., -1.7615455  -0.49462366\n",
      "  -1.31195211]\n",
      " [-2.07442877 -0.66946108 -1.72499303 ..., -1.32716049 -1.45041816\n",
      "  -1.86786402]]\n"
     ]
    }
   ],
   "source": [
    "coverage_matrix_vs_max_exon = coverage_matrix / exon_stats_array[:,-1]\n",
    "print coverage_matrix_vs_max_exon\n",
    "\n",
    "normalized_coverage_matrix = (coverage_matrix - exon_stats_array[:,1]) / exon_stats_array[:,2]\n",
    "print normalized_coverage_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get a matrix of the probability that any subject's reads will be for any particular exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['id' 'subject' 'specimen' ..., 'Ex77' 'Ex78' 'Ex79']\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L001' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 0.010921396708198739 0.0018458698661744347\n",
      "  0.012459621596677434]\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L002' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 0.011123296255679148 0.0021933260222465924\n",
      "  0.010496631677894407]\n",
      " ..., \n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L003' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 0.011693296438453986 0.0016959742926001963\n",
      "  0.012853699901812015]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L002' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 0.011130434782608696 0.0019130434782608696\n",
      "  0.01017391304347826]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L004' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 0.011900684931506849 0.0021404109589041095\n",
      "  0.010787671232876713]]\n"
     ]
    }
   ],
   "source": [
    "probability_in_an_exon_matrix = [list(coverage_matrix[0])]\n",
    "for row in coverage_matrix[1:]:\n",
    "    total_subj_reads = sum(map(int, row[len(base_headers):]))\n",
    "    new_row = list(row[:len(base_headers)])\n",
    "    new_row += [float(item) / total_subj_reads for item in row[len(base_headers):]]\n",
    "    probability_in_an_exon_matrix.append(new_row)\n",
    "probability_in_an_exon_matrix = np.array(probability_in_an_exon_matrix)\n",
    "print probability_in_an_exon_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
