{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/anaconda2/lib/python2.7/site-packages/setuptools-27.2.0-py2.7.egg/setuptools/dist.py:331: UserWarning: Normalizing '1.0.1.1-dev' to '1.0.1.1.dev0'\n",
      "running develop\n",
      "running egg_info\n",
      "writing requirements to genepeeks.egg-info/requires.txt\n",
      "writing genepeeks.egg-info/PKG-INFO\n",
      "writing top-level names to genepeeks.egg-info/top_level.txt\n",
      "writing dependency_links to genepeeks.egg-info/dependency_links.txt\n",
      "reading manifest file 'genepeeks.egg-info/SOURCES.txt'\n",
      "writing manifest file 'genepeeks.egg-info/SOURCES.txt'\n",
      "running build_ext\n",
      "Creating /home/ec2-user/anaconda2/lib/python2.7/site-packages/genepeeks.egg-link (link to .)\n",
      "genepeeks 1.0.1.1.dev0 is already the active version in easy-install.pth\n",
      "\n",
      "Installed /home/ec2-user/git/genepeeks-science\n",
      "Processing dependencies for genepeeks==1.0.1.1.dev0\n",
      "Searching for apiclient==1.0.3\n",
      "Best match: apiclient 1.0.3\n",
      "Adding apiclient 1.0.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for config==0.3.9\n",
      "Best match: config 0.3.9\n",
      "Adding config 0.3.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for dropbox==7.1.1\n",
      "Best match: dropbox 7.1.1\n",
      "Adding dropbox 7.1.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for ggplot==0.6.8\n",
      "Best match: ggplot 0.6.8\n",
      "Adding ggplot 0.6.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for GitPython==2.1.0\n",
      "Best match: GitPython 2.1.0\n",
      "Adding GitPython 2.1.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for google-api-python-client==1.5.5\n",
      "Best match: google-api-python-client 1.5.5\n",
      "Adding google-api-python-client 1.5.5 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for oauth2client==1.5.2\n",
      "Best match: oauth2client 1.5.2\n",
      "Adding oauth2client 1.5.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for mock==2.0.0\n",
      "Best match: mock 2.0.0\n",
      "Adding mock 2.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for nose==1.3.7\n",
      "Best match: nose 1.3.7\n",
      "Adding nose 1.3.7 to easy-install.pth file\n",
      "Installing nosetests script to /home/ec2-user/anaconda2/bin\n",
      "Installing nosetests-2.7 script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for weightedstats==0.3\n",
      "Best match: weightedstats 0.3\n",
      "Adding weightedstats 0.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for seaborn==0.7.1\n",
      "Best match: seaborn 0.7.1\n",
      "Adding seaborn 0.7.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for matplotlib==1.5.2\n",
      "Best match: matplotlib 1.5.2\n",
      "Adding matplotlib 1.5.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for multiprocess==0.70.4\n",
      "Best match: multiprocess 0.70.4\n",
      "Adding multiprocess 0.70.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for contextlib2==0.5.4\n",
      "Best match: contextlib2 0.5.4\n",
      "Adding contextlib2 0.5.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pytz==2015.7\n",
      "Best match: pytz 2015.7\n",
      "Adding pytz 2015.7 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pymongo==3.3.0\n",
      "Best match: pymongo 3.3.0\n",
      "Adding pymongo 3.3.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for scipy==0.18.0\n",
      "Best match: scipy 0.18.0\n",
      "Adding scipy 0.18.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for numpy==1.11.1\n",
      "Best match: numpy 1.11.1\n",
      "Adding numpy 1.11.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for mando==0.4\n",
      "Best match: mando 0.4\n",
      "Adding mando 0.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for commandwrapper==0.7\n",
      "Best match: commandwrapper 0.7\n",
      "Adding commandwrapper 0.7 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for httplib2==0.9.2\n",
      "Best match: httplib2 0.9.2\n",
      "Adding httplib2 0.9.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for boto==2.42.0\n",
      "Best match: boto 2.42.0\n",
      "Adding boto 2.42.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for urllib3==1.20\n",
      "Best match: urllib3 1.20\n",
      "Adding urllib3 1.20 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for requests==2.12.4\n",
      "Best match: requests 2.12.4\n",
      "Adding requests 2.12.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for typing==3.5.3.0\n",
      "Best match: typing 3.5.3.0\n",
      "Adding typing 3.5.3.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for six==1.10.0\n",
      "Best match: six 1.10.0\n",
      "Adding six 1.10.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for brewer2mpl==1.4.1\n",
      "Best match: brewer2mpl 1.4.1\n",
      "Adding brewer2mpl 1.4.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for statsmodels==0.6.1\n",
      "Best match: statsmodels 0.6.1\n",
      "Adding statsmodels 0.6.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for patsy==0.4.1\n",
      "Best match: patsy 0.4.1\n",
      "Adding patsy 0.4.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pandas==0.18.1\n",
      "Best match: pandas 0.18.1\n",
      "Adding pandas 0.18.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for gitdb2==2.0.0\n",
      "Best match: gitdb2 2.0.0\n",
      "Adding gitdb2 2.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for uritemplate==3.0.0\n",
      "Best match: uritemplate 3.0.0\n",
      "Adding uritemplate 3.0.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyasn1==0.1.9\n",
      "Best match: pyasn1 0.1.9\n",
      "Adding pyasn1 0.1.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyasn1-modules==0.0.8\n",
      "Best match: pyasn1-modules 0.0.8\n",
      "Adding pyasn1-modules 0.0.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for rsa==3.4.2\n",
      "Best match: rsa 3.4.2\n",
      "Adding rsa 3.4.2 to easy-install.pth file\n",
      "Installing pyrsa-encrypt-bigfile script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-encrypt script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-verify script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-sign script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-priv2pub script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-decrypt script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-decrypt-bigfile script to /home/ec2-user/anaconda2/bin\n",
      "Installing pyrsa-keygen script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pbr==1.10.0\n",
      "Best match: pbr 1.10.0\n",
      "Adding pbr 1.10.0 to easy-install.pth file\n",
      "Installing pbr script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for funcsigs==1.0.2\n",
      "Best match: funcsigs 1.0.2\n",
      "Adding funcsigs 1.0.2 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for python-dateutil==2.5.3\n",
      "Best match: python-dateutil 2.5.3\n",
      "Adding python-dateutil 2.5.3 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for pyparsing==2.1.4\n",
      "Best match: pyparsing 2.1.4\n",
      "Adding pyparsing 2.1.4 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for cycler==0.10.0\n",
      "Best match: cycler 0.10.0\n",
      "Adding cycler 0.10.0 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for dill==0.2.5\n",
      "Best match: dill 0.2.5\n",
      "Adding dill 0.2.5 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Sphinx==1.4.6\n",
      "Best match: Sphinx 1.4.6\n",
      "Processing Sphinx-1.4.6-py2.7.egg\n",
      "Sphinx 1.4.6 is already the active version in easy-install.pth\n",
      "Installing sphinx-apidoc script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-build script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-quickstart script to /home/ec2-user/anaconda2/bin\n",
      "Installing sphinx-autogen script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages/Sphinx-1.4.6-py2.7.egg\n",
      "Searching for smmap2==2.0.1\n",
      "Best match: smmap2 2.0.1\n",
      "Adding smmap2 2.0.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for imagesize==0.7.1\n",
      "Best match: imagesize 0.7.1\n",
      "Adding imagesize 0.7.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for alabaster==0.7.9\n",
      "Best match: alabaster 0.7.9\n",
      "Adding alabaster 0.7.9 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Babel==2.3.4\n",
      "Best match: Babel 2.3.4\n",
      "Adding Babel 2.3.4 to easy-install.pth file\n",
      "Installing pybabel script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for snowballstemmer==1.2.1\n",
      "Best match: snowballstemmer 1.2.1\n",
      "Adding snowballstemmer 1.2.1 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for docutils==0.12\n",
      "Best match: docutils 0.12\n",
      "Adding docutils 0.12 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Pygments==2.1.3\n",
      "Best match: Pygments 2.1.3\n",
      "Adding Pygments 2.1.3 to easy-install.pth file\n",
      "Installing pygmentize script to /home/ec2-user/anaconda2/bin\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for Jinja2==2.8\n",
      "Best match: Jinja2 2.8\n",
      "Adding Jinja2 2.8 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Searching for MarkupSafe==0.23\n",
      "Best match: MarkupSafe 0.23\n",
      "Adding MarkupSafe 0.23 to easy-install.pth file\n",
      "\n",
      "Using /home/ec2-user/anaconda2/lib/python2.7/site-packages\n",
      "Finished processing dependencies for genepeeks==1.0.1.1.dev0\n"
     ]
    }
   ],
   "source": [
    "! cd ../../genepeeks-science/ && python setup.py develop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from genepeeks.common import utilities as util\n",
    "import pysam, os\n",
    "import tabletext\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get DMD exons from gene collection in mongo using genepeeks science utilities function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished loading ensembl within the gene collection from the genepeeks-prod database with a single entries with a query of {'_id': 'DMD'} after 0.17 sec\n",
      "['Ex1', 'Ex2', 'Ex3', 'Ex4', 'Ex5', 'Ex6', 'Ex7', 'Ex8', 'Ex9', 'Ex10', 'Ex11', 'Ex12', 'Ex13', 'Ex14', 'Ex15', 'Ex16', 'Ex17', 'Ex18', 'Ex19', 'Ex20', 'Ex21', 'Ex22', 'Ex23', 'Ex24', 'Ex25', 'Ex26', 'Ex27', 'Ex28', 'Ex29', 'Ex30', 'Ex31', 'Ex32', 'Ex33', 'Ex34', 'Ex35', 'Ex36', 'Ex37', 'Ex38', 'Ex39', 'Ex40', 'Ex41', 'Ex42', 'Ex43', 'Ex44', 'Ex45', 'Ex46', 'Ex47', 'Ex48', 'Ex49', 'Ex50', 'Ex51', 'Ex52', 'Ex53', 'Ex54', 'Ex55', 'Ex56', 'Ex57', 'Ex58', 'Ex59', 'Ex60', 'Ex61', 'Ex62', 'Ex63', 'Ex64', 'Ex65|66', 'Ex67', 'Ex68', 'Ex69', 'Ex70', 'Ex71', 'Ex72', 'Ex73', 'Ex74', 'Ex75', 'Ex76', 'Ex77', 'Ex78', 'Ex79']\n"
     ]
    }
   ],
   "source": [
    "DMD_ensembl = util.Mongo.get_collection_data('gene', wanted_db='prod', query={'_id': 'DMD'}, find_one=True, single_field='ensembl')\n",
    "DMD_exons = util.get_nested_value(DMD_ensembl, ('is_primary', 'transcripts', 'is_primary', 'exons'))\n",
    "DMD_exons_merged = util.merge_intervals(DMD_exons, min_dist=200, include_index=True)\n",
    "exon_labels = ['Ex' + exon['index'] for exon in DMD_exons_merged]\n",
    "print exon_labels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Count the number of males and females"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180\n",
      "263\n"
     ]
    }
   ],
   "source": [
    "F_count = M_count = 0\n",
    "for root, dirs, files in os.walk('/mnt/vep/subjects'):\n",
    "    for file_name in files:\n",
    "        if file_name.endswith('.bam'):\n",
    "            if file_name.startswith('F'):\n",
    "                F_count += 1\n",
    "            elif file_name.startswith('M'):\n",
    "                M_count += 1\n",
    "print F_count\n",
    "print M_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get each sample's coverage across all DMD exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished parsing 10 subjects\n",
      "Finished parsing 20 subjects\n",
      "Finished parsing 30 subjects\n",
      "Finished parsing 40 subjects\n",
      "Finished parsing 50 subjects\n",
      "Finished parsing 60 subjects\n",
      "Finished parsing 70 subjects\n",
      "Finished parsing 80 subjects\n",
      "Finished parsing 90 subjects\n",
      "Finished parsing 100 subjects\n",
      "Finished parsing 110 subjects\n",
      "Finished parsing 120 subjects\n",
      "Finished parsing 130 subjects\n",
      "Finished parsing 140 subjects\n",
      "Finished parsing 150 subjects\n",
      "Finished parsing 160 subjects\n",
      "FK2R-0000-151I.bam is missing an index file\n",
      "Finished parsing 170 subjects\n",
      "Finished parsing 180 subjects\n",
      "Finished parsing 190 subjects\n",
      "Finished parsing 200 subjects\n",
      "Finished parsing 210 subjects\n",
      "Finished parsing 220 subjects\n",
      "Finished parsing 230 subjects\n",
      "Finished parsing 240 subjects\n",
      "Finished parsing 250 subjects\n",
      "Finished parsing 260 subjects\n",
      "Finished parsing 270 subjects\n",
      "Finished parsing 280 subjects\n",
      "Finished parsing 290 subjects\n",
      "Finished parsing 300 subjects\n",
      "Finished parsing 310 subjects\n",
      "Finished parsing 320 subjects\n",
      "Finished parsing 330 subjects\n",
      "Finished parsing 340 subjects\n",
      "Finished parsing 350 subjects\n",
      "Finished parsing 360 subjects\n",
      "Finished parsing 370 subjects\n",
      "Finished parsing 380 subjects\n",
      "Finished parsing 390 subjects\n",
      "Finished parsing 400 subjects\n",
      "Finished parsing 410 subjects\n",
      "Finished parsing 420 subjects\n",
      "Finished parsing 430 subjects\n",
      "MCLR-NA12-8911.bam is missing an index file\n",
      "Finished parsing 440 subjects\n",
      "441\n",
      "[['id' 'subject' 'specimen' ..., 'Ex77' 'Ex78' 'Ex79']\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L001' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 71 12 81]\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L002' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 71 14 67]\n",
      " ..., \n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L003' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 131 19 144]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L002' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 128 22 117]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L004' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 139 25 126]]\n"
     ]
    }
   ],
   "source": [
    "base_headers = ['id', 'subject', 'specimen', 'sample', 'gender', 'sequencer', 'flow_cell_id', 'lane', 'bwa_version', 'date_modified']\n",
    "subject_count = 0\n",
    "coverage_matrix = [base_headers + exon_labels]\n",
    "for root, dirs, files in os.walk('/mnt/vep/subjects'):\n",
    "    for file_name in files:\n",
    "        if file_name.endswith('.bam'):\n",
    "            if '{}.bai'.format(file_name) not in files:\n",
    "                print '{} is missing an index file'.format(file_name)\n",
    "                continue\n",
    "            file_path = os.path.join(root, file_name)\n",
    "            date_modified = os.path.getmtime(file_path)\n",
    "            \n",
    "            bamfile = pysam.AlignmentFile(file_path, \"rb\")\n",
    "            \n",
    "            # Gather identifying info for each sample\n",
    "            subject_coverages = {}\n",
    "            bwa_version = next(PG['VN'] for PG in bamfile.header['PG'] if PG.get('ID') == 'bwa')\n",
    "            for RG in bamfile.header['RG']:\n",
    "                try:\n",
    "                    # normal RG['ID'] format: FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L004\n",
    "                    subject, specimen_sample, flow_cell_lane = RG['ID'].split('_')\n",
    "                except:\n",
    "                    # older RG['ID'] format: FPWB-0000-429L_1-P1-1\n",
    "                    subject, specimen_sample = RG['ID'].split('_')\n",
    "                    flow_cell_id = lane = None\n",
    "                else:\n",
    "                    flow_cell_id, lane = flow_cell_lane.rsplit('-', 1)\n",
    "\n",
    "                gender = subject[0]\n",
    "                if specimen_sample.startswith(('ACGT', 'Omega')):\n",
    "                    lab, specimen_num, sequencer, sample = specimen_sample.split('.')\n",
    "                    specimen_num = '{}_{}'.format(lab, specimen_num)\n",
    "                else:\n",
    "                    specimen_num, sequencer, sample = specimen_sample.split('-')\n",
    "                specimen = '{}_{}'.format(subject, specimen_num)\n",
    "                sample = '{}_{}'.format(subject, specimen_sample)\n",
    "                full_id = RG['ID']\n",
    "                if root.endswith('re86'):\n",
    "                    full_id += '_re86'\n",
    "                    \n",
    "                id_info = [full_id, subject, specimen, sample, gender, sequencer, flow_cell_id, lane, bwa_version, date_modified]\n",
    "                # Initialize each row with identifying info for the sample plus each exon's coverage of 0\n",
    "                subject_coverages[RG['ID']] = id_info + [0] * len(DMD_exons_merged)\n",
    "\n",
    "            # Get coverage data for each sample within each exon\n",
    "            for read in bamfile.fetch('X', start=31137345, end=33229636):\n",
    "                if not read.is_unmapped and read.mapping_quality == 60:\n",
    "                    # Find what exon each read falls in, and increase that exon's coverage by 1\n",
    "                    interval_info = util.in_interval(read.reference_start, DMD_exons_merged, get_interval=True)\n",
    "                    if not interval_info[0]:\n",
    "                        # If the start of the read is not in an exon, check the end of the read\n",
    "                        interval_info = util.in_interval(read.reference_end, DMD_exons_merged, get_interval=True)\n",
    "#                     else:\n",
    "#                         # Make sure that the end of the read does not map to a different exon than the start\n",
    "#                         end_interval_info = util.in_interval(read.reference_end, DMD_exons_merged, get_interval=True)\n",
    "#                         if end_interval_info[0] and end_interval_info[1] != interval_info[1]:\n",
    "#                             print read.reference_start, interval_info[1], read.reference_end, end_interval_info[1]\n",
    "                        \n",
    "                    if interval_info[0]:\n",
    "                        exon_num = interval_info[1]\n",
    "                        subject_coverages[read.get_tag('RG')][exon_num + len(base_headers)] += 1\n",
    "                        \n",
    "            coverage_matrix += subject_coverages.values()\n",
    "            subject_count += 1\n",
    "            if subject_count % 10 == 0:\n",
    "                print 'Finished parsing {} subjects'.format(subject_count)\n",
    "coverage_matrix = np.array(coverage_matrix)\n",
    "print subject_count\n",
    "print coverage_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.DataFrame(coverage_matrix)\n",
    "df.to_csv(\"../exon_data/coverage_matrix.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Print each exon's mean, std, min, and max coverage, and turn the stats into an array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1267 rows\n",
      "┌─────────┬────────┬────────┬─────┬───────┬────────┬──────┐\n",
      "│ Exon    │ Mean   │ Std    │ Min │ 1%    │ 99%    │ Max  │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex1     │ 440.51 │ 336.99 │   0 │ 39.46 │ 2051.7 │ 2647 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex2     │  15.19 │   9.32 │   0 │   0.0 │   39.0 │   42 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex3     │  106.5 │  56.55 │   0 │ 24.92 │  287.1 │  476 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex4     │ 179.87 │ 118.86 │   0 │ 41.46 │ 720.26 │  982 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex5     │ 113.31 │  66.95 │   0 │ 29.46 │  411.4 │  552 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex6     │ 155.07 │  86.83 │   0 │ 25.46 │ 476.54 │  654 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex7     │ 130.78 │  86.01 │   0 │  4.46 │ 328.48 │  461 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex8     │  96.99 │  62.92 │   0 │  1.92 │  257.2 │  370 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex9     │  35.18 │  20.01 │   0 │   3.0 │  91.08 │  110 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex10    │  135.0 │  83.56 │   0 │  18.3 │ 405.56 │  535 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex11    │ 118.08 │  48.64 │   0 │  32.0 │ 280.54 │  366 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex12    │ 106.17 │  48.69 │   0 │  33.0 │ 282.68 │  385 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex13    │ 110.15 │  54.82 │   0 │ 32.46 │ 314.64 │  407 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex14    │  57.65 │  30.34 │   0 │   6.0 │ 120.54 │  144 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex15    │ 159.45 │  74.36 │   0 │ 31.46 │ 350.18 │  443 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex16    │  70.93 │  41.97 │   0 │ 14.46 │ 247.94 │  323 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex17    │  87.18 │  49.54 │   0 │  7.46 │  223.4 │  283 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex18    │   51.9 │  24.82 │   0 │  10.0 │ 116.54 │  166 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex19    │  89.85 │  58.52 │   0 │  4.46 │ 273.16 │  344 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex20    │  257.2 │ 128.91 │   0 │ 61.76 │ 612.26 │  818 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex21    │ 125.86 │  68.24 │   0 │ 23.46 │  389.7 │  544 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex22    │ 162.81 │  94.23 │   0 │  44.0 │ 608.26 │  685 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex23    │  131.6 │  60.93 │   0 │  35.3 │ 320.94 │  434 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex24    │  95.36 │  73.15 │   0 │  9.38 │ 456.92 │  545 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex25    │ 234.19 │ 160.27 │   0 │ 20.92 │ 770.02 │ 1026 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex26    │ 102.91 │  76.74 │   0 │ 16.92 │ 470.12 │  608 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex27    │  189.1 │ 144.37 │   0 │ 24.92 │  870.0 │ 1099 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex28    │  78.71 │  43.51 │   0 │  9.38 │ 185.62 │  260 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex29    │ 181.59 │  88.77 │   0 │ 61.92 │  510.4 │  674 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex30    │  62.64 │  37.33 │   0 │  8.38 │ 197.64 │  304 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex31    │  77.36 │  45.01 │   0 │  4.46 │ 173.08 │  201 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex32    │ 122.04 │ 104.34 │   0 │ 28.92 │  691.2 │  847 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex33    │ 186.73 │  93.52 │   0 │ 42.46 │ 499.18 │  666 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex34    │  97.22 │  61.33 │   0 │ 25.46 │  384.5 │  473 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex35    │ 127.08 │  60.79 │   0 │ 36.84 │ 351.02 │  406 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex36    │  124.4 │  60.26 │   0 │ 36.38 │ 356.24 │  452 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex37    │  138.2 │  80.39 │   0 │ 34.84 │ 480.34 │  651 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex38    │  81.53 │  44.81 │   0 │  13.3 │ 255.78 │  339 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex39    │ 123.49 │  54.75 │   0 │ 44.38 │  318.0 │  351 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex40    │  129.5 │  57.94 │   0 │ 26.92 │  320.7 │  411 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex41    │ 116.06 │  59.69 │   0 │ 34.84 │ 332.34 │  452 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex42    │  89.56 │  57.35 │   0 │ 20.46 │ 368.72 │  443 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex43    │ 167.88 │   72.2 │   0 │ 46.76 │  405.7 │  628 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex44    │ 119.63 │  90.41 │   0 │ 10.46 │ 477.22 │  592 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex45    │ 228.95 │ 153.47 │   0 │ 17.84 │ 634.28 │  826 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex46    │  102.7 │  65.49 │   0 │ 23.84 │ 387.08 │  533 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex47    │  195.9 │ 135.36 │   0 │  18.0 │ 677.36 │  873 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex48    │ 134.49 │  72.74 │   0 │ 38.46 │ 457.26 │  555 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex49    │ 161.52 │ 104.36 │   0 │  9.38 │ 387.08 │  469 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex50    │  90.04 │  47.31 │   0 │ 19.46 │  255.2 │  347 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex51    │ 117.69 │   59.5 │   0 │ 34.84 │ 316.72 │  465 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex52    │  82.92 │  46.18 │   0 │ 24.84 │ 273.08 │  345 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex53    │ 181.24 │ 101.02 │   0 │ 38.46 │ 593.98 │  708 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex54    │   87.6 │  48.72 │   0 │ 16.84 │ 252.16 │  384 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex55    │ 204.75 │ 106.14 │   0 │ 51.84 │ 611.68 │  761 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex56    │  77.81 │  36.94 │   0 │  20.0 │ 229.64 │  247 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex57    │  82.48 │  51.76 │   0 │ 19.38 │ 316.26 │  478 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex58    │ 104.25 │  53.93 │   0 │ 30.38 │ 289.08 │  438 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex59    │ 140.07 │  61.15 │   0 │  41.0 │ 378.08 │  426 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex60    │ 194.99 │ 145.94 │   0 │ 22.84 │  843.8 │ 1095 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex61    │  78.64 │  51.44 │   0 │ 14.46 │  312.7 │  387 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex62    │ 148.86 │  89.24 │   0 │ 18.22 │ 473.48 │  586 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex63    │   86.5 │  62.02 │   0 │  21.0 │  385.8 │  496 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex64    │ 113.32 │  62.18 │   0 │ 18.46 │ 249.86 │  359 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex65|66 │ 204.26 │ 136.32 │   0 │ 48.92 │ 832.02 │  996 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex67    │ 127.75 │  66.65 │   0 │ 30.22 │ 368.96 │  476 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex68    │ 149.05 │  89.16 │   0 │  13.3 │ 356.24 │  359 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex69    │ 112.12 │  74.01 │   0 │ 28.46 │ 482.56 │  566 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex70    │  139.5 │  63.59 │   0 │ 46.46 │ 334.24 │  527 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex71    │  53.35 │  23.52 │   0 │ 10.46 │ 122.86 │  160 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex72    │ 115.36 │  63.63 │   0 │ 13.84 │ 344.02 │  408 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex73    │  68.23 │  35.77 │   0 │   9.0 │ 154.54 │  222 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex74    │  177.1 │  93.57 │   0 │ 56.46 │ 574.24 │  720 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex75    │  56.53 │   42.3 │   0 │   6.0 │ 254.24 │  298 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex76    │  47.48 │  22.11 │   0 │  10.0 │ 120.08 │  149 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex77    │ 110.17 │  65.91 │   0 │ 20.46 │ 410.96 │  482 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex78    │  21.43 │  19.38 │   0 │  0.46 │  112.4 │  152 │\n",
      "├─────────┼────────┼────────┼─────┼───────┼────────┼──────┤\n",
      "│ Ex79    │ 103.09 │  59.52 │   0 │  18.0 │  302.2 │  399 │\n",
      "└─────────┴────────┴────────┴─────┴───────┴────────┴──────┘\n"
     ]
    }
   ],
   "source": [
    "print len(coverage_matrix), 'rows'\n",
    "exon_cov_stats = [['Exon', 'Mean', 'Std', 'Min', '1%', '99%', 'Max']]\n",
    "exon_max_cov = []\n",
    "female_coverage_matrix = np.array([list(coverage_matrix[0])] + [list(item) for item in coverage_matrix if item[0].startswith('F')])\n",
    "for i in range(len(DMD_exons_merged)):\n",
    "    col = i + len(base_headers)\n",
    "    exon_coverage = map(int, female_coverage_matrix[1:,col])\n",
    "    exon_stats = [\n",
    "        coverage_matrix[0][col],\n",
    "        round(np.mean(exon_coverage), 2),\n",
    "        round(np.std(exon_coverage), 2),\n",
    "        min(exon_coverage),\n",
    "        np.percentile(exon_coverage, 1),\n",
    "        np.percentile(exon_coverage, 99),\n",
    "        max(exon_coverage)\n",
    "    ]\n",
    "    exon_cov_stats.append(exon_stats)\n",
    "# exon_stats_array = np.array(exon_cov_stats[1:,[1:5]])\n",
    "print tabletext.to_text(exon_cov_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Normalize each subject's coverage on each exon compared to the mean and std of that exon's coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.67197452  0.34883721  0.56981132 ...,  0.7337884   0.79245283\n",
      "   0.73668639]\n",
      " [ 0.65059145  0.41860465  0.55471698 ...,  0.6894198   0.49056604\n",
      "   0.73964497]\n",
      " [ 0.78207461  0.93023256  0.68301887 ...,  0.83959044  0.73584906\n",
      "   0.74556213]\n",
      " ..., \n",
      " [ 0.58826206  0.62790698  0.57735849 ...,  0.47098976  0.73584906\n",
      "   0.5443787 ]\n",
      " [ 0.51319381  0.37209302  0.35471698 ...,  0.49829352  0.58490566\n",
      "   0.53846154]\n",
      " [ 0.39126479  0.44186047  0.40754717 ...,  0.56313993  0.43396226\n",
      "   0.46153846]]\n",
      "[[-0.09158981 -1.14850299 -0.52688771 ..., -0.18404207  0.81959379\n",
      "   0.12059012]\n",
      " [-0.24263264 -0.78922156 -0.63833937 ..., -0.48125286 -1.09199522\n",
      "   0.14197135]\n",
      " [ 0.68612013  1.84550898  0.30899972 ...,  0.52469136  0.46117085\n",
      "   0.1847338 ]\n",
      " ..., \n",
      " [-0.68290645  0.28862275 -0.47116188 ..., -1.94444444  0.46117085\n",
      "  -1.26918965]\n",
      " [-1.21316322 -1.02874251 -2.11507384 ..., -1.7615455  -0.49462366\n",
      "  -1.31195211]\n",
      " [-2.07442877 -0.66946108 -1.72499303 ..., -1.32716049 -1.45041816\n",
      "  -1.86786402]]\n"
     ]
    }
   ],
   "source": [
    "coverage_matrix_vs_max_exon = coverage_matrix / exon_stats_array[:,-1]\n",
    "print coverage_matrix_vs_max_exon\n",
    "\n",
    "normalized_coverage_matrix = (coverage_matrix - exon_stats_array[:,1]) / exon_stats_array[:,2]\n",
    "print normalized_coverage_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get a matrix of the probability that any subject's reads will be for any particular exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['id' 'subject' 'specimen' ..., 'Ex77' 'Ex78' 'Ex79']\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L001' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 0.010921396708198739 0.0018458698661744347\n",
      "  0.012459621596677434]\n",
      " ['FCLR-GP01-2121_1-M1-1_HGGF5AFXX-L002' 'FCLR-GP01-2121'\n",
      "  'FCLR-GP01-2121_1' ..., 0.011123296255679148 0.0021933260222465924\n",
      "  0.010496631677894407]\n",
      " ..., \n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L003' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 0.011693296438453986 0.0016959742926001963\n",
      "  0.012853699901812015]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L002' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 0.011130434782608696 0.0019130434782608696\n",
      "  0.01017391304347826]\n",
      " ['FPWB-0000-952M_1-M1-1_H753VAFXX-L004' 'FPWB-0000-952M'\n",
      "  'FPWB-0000-952M_1' ..., 0.011900684931506849 0.0021404109589041095\n",
      "  0.010787671232876713]]\n"
     ]
    }
   ],
   "source": [
    "probability_in_an_exon_matrix = [list(coverage_matrix[0])]\n",
    "for row in coverage_matrix[1:]:\n",
    "    total_subj_reads = sum(map(int, row[len(base_headers):]))\n",
    "    new_row = list(row[:len(base_headers)])\n",
    "    new_row += [float(item) / total_subj_reads for item in row[len(base_headers):]]\n",
    "    probability_in_an_exon_matrix.append(new_row)\n",
    "probability_in_an_exon_matrix = np.array(probability_in_an_exon_matrix)\n",
    "print probability_in_an_exon_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
