{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment Statistics and Deletion Simulation with pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pysam\n",
    "import numpy as np\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from genepeeks.common import utilities as util\n",
    "import seaborn as sns\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished loading the gene collection from the genepeeks-prod database with a single entries with a query of {'_id': 'DMD'} after 0.32 sec\n"
     ]
    }
   ],
   "source": [
    "DMD_ensembl = util.Mongo.get_collection_data('gene', wanted_db='prod', query={'_id': 'DMD'}, find_one=True, single_field='ensembl')\n",
    "DMD_exons = util.get_nested_value(DMD_ensembl, ('is_primary', 'transcripts', 'is_primary', 'exons'))\n",
    "# had to make local change in utilities \n",
    "DMD_exons_merged = util.merge_intervals(DMD_exons, min_dist=200, include_index=True)\n",
    "exon_labels = ['Ex' + exon['index'] for exon in DMD_exons_merged]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2702\n"
     ]
    }
   ],
   "source": [
    "# DMD exon/gene coordinates and useful regions (coordinates taken from IGV b37 reference)\n",
    "EXON46 = [31950197, 31950344]\n",
    "EXON47 = [31947713, 31947862]\n",
    "EXON48 = [31893308, 31893490]\n",
    "EXON49 = [31854835, 31854939]\n",
    "EXON50 = [31838092, 31838200]\n",
    "EXON51 = [31792077, 31792309]\n",
    "DMD = [31115794, 33357558]  # taken from ensembl GRCh37\n",
    "\n",
    "# deletion of partial exons\n",
    "EX48_50_PART = [31838130, 31893308]\n",
    "\n",
    "# for consistency with exon labeling in Ensembl\n",
    "exon_dict = {exon['index']: [exon['start'], exon['end']] for exon in DMD_exons_merged}\n",
    "print exon_dict['1'][1] - exon_dict['1'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Velina's local bam files for testing (names are consistent with subject/sample names)\n",
    "bamfiles = ['FGPC-0000-444F.bam', 'FGPC-8V7M-82J3.bam', 'FPWB-0000-840T.bam', 'MGPC-6KHG-N0LS.bam']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# workaround to force real time print output in notebooks\n",
    "# see http://stackoverflow.com/questions/29772158/make-ipython-notebook-print-in-real-time\n",
    "import sys\n",
    "oldsysstdout = sys.stdout\n",
    "class flushfile():\n",
    "    def __init__(self, f):\n",
    "        self.f = f\n",
    "    def __getattr__(self,name): \n",
    "        return object.__getattribute__(self.f, name)\n",
    "    def write(self, x):\n",
    "        self.f.write(x)\n",
    "        self.f.flush()\n",
    "    def flush(self):\n",
    "        self.f.flush()\n",
    "sys.stdout = flushfile(sys.stdout)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Some exploration with exon length/coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<matplotlib.text.Text at 0x10592ae90>, <matplotlib.text.Text at 0x10d5f42d0>)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh4AAAF9CAYAAABLS2EjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAAPYQAAD2EBqD+naQAAIABJREFUeJzt3XmYXHWd7/F3p2OgIWmIUcEtCopfw+IYlEVAQR2vMnNH\nHS7jOnqRcUPcHfVxRHFEXAARxQUX1EEHrvuKzqi4IKOMMG4sw9eMbLIqMdBJCIR09/3jVJOi01m6\nu+p3Tle9X8+TJ13nnD7nW9/8uvqT3zl1amB8fBxJkqQS5tVdgCRJ6h8GD0mSVIzBQ5IkFWPwkCRJ\nxRg8JElSMQYPSZJUjMFDkiQVY/CQJEnFGDwkSVIx8+suoF1EbAdcDBybmee3lj0e+ADwSOB3wBsz\n87z6qpQkSTPVmBmPVug4B9izbdl9gW8CZwN7A18CvhERD6ilSEmSNCuNCB4RsQy4ENht0qqDgbsy\n89TMvDoz3wPcARxYukZJkjR7jQgewKHAecDjgIG25SuBJRHxtwAR8UxgIXBJ8QolSdKsDTTt02kj\nYgw4rO0aj9OBVwBjVEHpRZl5Vo0lSpKkGWrKjMeUImIhsDvwdmA/4ETg9Ih4RK2FSZKkGWnUu1qm\n8GaAzDyx9fjXEXEg8Brg2G3Zwfj4+PjAwMDWN5QkSZN1/Bdo04PHvsBvJi37FbDXtu5gYGCAkZF1\njI6OdbSwuWZwcB7Dw0N93wv7ULEPG9mLin3YyF5UJvrQaU0PHjfQ9vbalkcCV01nJ6OjY2zY0L+D\np529qNiHin3YyF5U7MNG9qI7mh48PgX8NCJeQ3U/j2cATwUeXWtVkiRpRpp4cendb7PJzP8EjgCO\nojrl8nzg8My8op7SJEnSbDRuxiMzByc9/jbw7ZrKkSRJHdTEGQ9JktSjDB6SJKkYg4ckSSrG4CFJ\nkooxeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKkYgwekiSpGIOHJEkqpnGf1SKp+9avX89FF13O\nyMg6Rkfr+djvvfbahwULFtRybEn1MXhIfejSSy/hDSd/mUVLltZy/NUrr+Wk18Py5Y+p5fiS6mPw\nkPrUoiVL2XnXPeouQ1Kf8RoPSZJUjMFDkiQVY/CQJEnFGDwkSVIxBg9JklSMwUOSJBVj8JAkScUY\nPCRJUjGNuoFYRGwHXAwcm5nnt5Y9GPg4cChwPfDWzPxSfVVKkqSZasyMRyt0nAPs2bZsEPgOcAfw\naOAU4PMRseeUO5EkSY3WiBmPiFgGnD3Fqr8GHggcmJlrgRUR8TTgIODygiVKkqQOaETwoDqNch5w\nHHD75OWt0AFAZh5RuDZJktQhjQgemXnGxNcR0b5qd+CqiHgP8ALgT8A7MvMbZSuUJEmd0JhrPDZj\nIfAiYGfgfwOfA74cEfvWWpUkSZqRRsx4bMEG4JbMPKb1+NcR8XjgpcDLt3Ung4NNz1fdN9GDfu+F\nfajMmzdQdwkMDs5j/vz6/x0cExX7sJG9qHTr+Tc9eNwIjE1alsA+09nJ8PBQxwqa6+xFpd/7sHDh\n9nWXwPDwEIsX71h3GXfr9zExwT5sZC+6o+nB40LgrRExkJnjrWXLgKuns5ORkXWMjk7OL/1lcHAe\nw8NDfd8L+1BZs+aOuktgZGQdq1at3fqGXeaYqNiHjexFZaIPndb04HEO8DbgoxFxCvBU4GnA/tPZ\nyejoGBs29O/gaWcvKv3eh7Gx8a1v1GVN+zdoWj11sQ8b2YvuaOIJrLtfETNzNfAUqlmOS4BXAc/K\nzN/UVJskSZqFxs14ZObgpMdXAIfVU40kSeqkJs54SJKkHmXwkCRJxRg8JElSMQYPSZJUjMFDkiQV\nY/CQJEnFGDwkSVIxBg9JklSMwUOSJBVj8JAkScUYPCRJUjEGD0mSVIzBQ5IkFWPwkCRJxRg8JElS\nMQYPSZJUjMFDkiQVY/CQJEnFGDwkSVIxBg9JklSMwUOSJBVj8JAkScUYPCRJUjGNCh4RsV1EXBIR\nT5hi3XBEXB8RL6yjNkmSNHuNCR4RsR1wDrDnZjY5Cdi1XEWSJKnTGhE8ImIZcCGw22bWHwI8Cbip\nZF2SJKmzGhE8gEOB84DHAQPtKyLiXsDHgVcA68uXJkmSOmV+3QUAZOYZE19HxOTVxwG/zMwfTLFO\nkiTNIY0IHpsTEXsCLwX2mc1+BgebMrFTn4ke9Hsv7ENl3ryBrW/UZYOD85g/v/5/B8dExT5sZC8q\n3Xr+jQ4ewCeAt2fmLbPZyfDwUIfKmfvsRaXf+7Bw4fZ1l8Dw8BCLF+9Ydxl36/cxMcE+bGQvuqOx\nwSMilgIHAY+KiFNbi3cAzoiIZ2fmX2/rvkZG1jE6OtaNMueMwcF5DA8P9X0v7ENlzZo76i6BkZF1\nrFq1tu4yHBMt9mEje1GZ6EOnNTZ4ANcBD5+07CfAacDZ09nR6OgYGzb07+BpZy8q/d6HsbHxukto\n3L9B0+qpi33YyF50R2ODR2aOAVe2L4uIDcCfMvPGeqqSJEmz0cQrZ7b0X7H6/5smSZJmrHEzHpk5\nuIV1u5esRZIkdVYTZzwkSVKPMnhIkqRiDB6SJKkYg4ckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJ\nKsbgIUmSijF4SJKkYhp3y3RJvW9sdAOZV9R2/L322ocFCxbUdnypnxk8JBW39tYbOfPcG1h04Zri\nx1698lpOej0sX/6Y4seWZPCQVJNFS5ay86571F2GpMK8xkOSJBVj8JAkScUYPCRJUjEGD0mSVIzB\nQ5IkFWPwkCRJxRg8JElSMQYPSZJUjMFDkiQV06g7l0bEdsDFwLGZeX5r2YHA+4FHAdcBp2TmmfVV\nKUmSZqoxMx6t0HEOsGfbsl2A7wA/BB4NvAM4PSIOr6NGSZI0O42Y8YiIZcDZU6x6JnBjZr6t9fj3\nEfFE4HnAd0vVJ0mSOqMpMx6HAucBjwMG2pZ/F3jRFNvvVKIoSZLUWY2Y8cjMMya+joj25dcC17at\nux/wHODtJeuTJEmd0ZQZj62KiO2BrwA3AJ+ouRxJkjQDjZjx2JqI2BH4JvBw4ODMvGM63z84OGfy\nVddM9KDfe2EfKvPmDWx9ox42ODiP+fPvORb6fUzYh43sRaVbz7/xwSMiFgH/BuwOPDEzr5zuPoaH\nhzpe11xlLyr93oeFC7evu4RaDQ8PsXjxjpssk31oZy+6o9HBIyIGgK8BDwWekJkrZrKfkZF1jI6O\ndbK0OWdwcB7Dw0N93wv7UFmzZlqThj1nZGQdq1atBRwTE+zDRvaiMtGHTmt08ABeDBwG/A0w0rqv\nB8D6zFy1rTsZHR1jw4b+HTzt7EWl3/swNjZedwm1murfv9/HxAT7sJG96I4mBo/x1h+AI6jeXvvt\nSdv8BHhSyaIkSdLsNS54ZOZg29feoVSSpB7S35fsSpKkogwekiSpGIOHJEkqxuAhSZKKMXhIkqRi\nDB6SJKkYg4ckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJKqZxn9UilbJ+/Xouu+yS2o5/1113AXCv\ne92r+LFXrMjix5QkMHioj1122SW86dSvsmjJ0lqOf/OVF7HDTrvUcvybr7yIXXbfr/hxJcngob62\naMlSdt51j1qOvXrlH1i05MG1HH/1yj8UP6Ykgdd4SJKkggwekiSpGIOHJEkqxuAhSZKKMXhIkqRi\nDB6SJKkYg4ckSSrG4CFJkooxeEiSpGIadefSiNgOuBg4NjPPby17KPBJ4HHA1cDrMvP7ddUoSZJm\nrjEzHq3QcQ6w56RVXwduAB4DfB74WkQ8qHB5kiSpAxoRPCJiGXAhsNuk5U8CdgdelpX3Aj8Hji5f\npSRJmq1GBA/gUOA8qtMpA23LDwB+mZl3tC27oLWdJEmaYxpxjUdmnjHxdUS0r7o/1WmWdjcDnmqR\nJGkOmlHwiIgXAl/IzDsnLd8ReGlmfqATxQE7AHdOWnYnsN10djI42JSJnfpM9KDfe9Heh37vRT8b\nHJzH/Pn3/Jno9/FgHzayF5VuPf9tDh4RcR+qIADwGeDSiLhl0mbLgXcDnQoedwD3nrRsO+D26exk\neHioQ+XMffaiMjw8ZC/62PDwEIsX77jJMtmHdvaiO6Yz4/FXwGeBcarrMC6aYpsB4DuzL+tu17Pp\nu1x2BW6czk5GRtYxOjrWsaLmosHBeQwPD/V9L9r7MDKyru5yVJORkXWsWrUW8Gdjgn3YyF5UJvrQ\nadscPDLzrIi4muqC1B8CRwJ/bttkHFgDXNLB+i4E3hwR27Wd1jkE+Ol0djI6OsaGDf07eNrZi8ro\n6Fhfv6D0u6l+DvzZqNiHjexFd0zrGo+2m3o9EfiPzNzQlao2+gnwB+CzEXEC8HRgP+CoLh9XkiR1\nwYwuLs3Mn0TEIRFxMLCAe74Flsx85yxqGm/bz1hEPAM4k+qOpv8DPDMzr5vF/iVJUk1m+q6W44B3\nArcCt01aPd5aNyOZOTjp8ZXAE2e6P0mS1BwzvY/HK4C3ZuZ7OlmMJEnqbTN9k+7OwNmdLESSJPW+\nmQaP/wAO6mQhkiSp9830VMvZwIcj4rHAFUy6u2hmnjXbwiRJUu+ZafA4s/X366ZYNw4YPCRJ0iZm\n+nba/r6BvSRJmhEDhCRJKmam9/G4irYbfU2WmbvPuCJJktSzZnqNx79wz+AxH3gE8DTg7bMtSpIk\n9aaZXuPxjqmWR8TLgL8EPjiLmiRJUo/q9DUe/wYc3uF9SpKkHtHp4HEkMNLhfUqSpB7RyYtLFwH3\nBo6fbVGSJKk3deriUoD1wM8z88ezqkiSJPWsjl5cKkmStCUznfEgIh4DvBHYB7gLuAw4LTMv6lBt\nkiSpx8zo4tKIOBT4GfBw4HvAT4BHAhdExMGdK0+SJPWSmc54nAh8OjOPaV8YER8B3gU8cbaFSZKk\n3jPT4LEv8OIplp8OeKpFkiRNaab38bgFuM8Uy+8H3DnzciRJUi+bafD4FvDhiFg2sSAi9gQ+1Fon\nSZK0iZmeajkO+D5waUTcRnVPj8XAr4F/7FBtkiSpx0w7eETEDsCtwP7AU4G9gSGqazv+PTPHOllg\nRDwI+BjwBGAl8MHM9EPoJEmag6Z1qiUingtcDeybmWOZ+d3MPBk4CPhX4BmdL5EvAaupLmh9LXBi\nRHTjOJIkqcu2OXhExGHA56iu4bh+0urXAt8AvhgRB3WquIjYGTgAeFdm/j4zv0n1CbhP7tQxJElS\nOdM51fIW4PTMfN3kFZl5BfCiiBinuv7jrzpU3zpgbWvfbwEeBhzcqkWSJM0x0znVshw4cyvbfJTq\nlEhHZOadwCuBl1OFkP8GvpOZn+3UMSRJUjnTCR7bU/3y35I/AzvMvJwpLQO+SXUx61HAka1rTSRJ\n0hwznVMtCTwO+P0WtjkIuGZWFbWJiCcD/wA8qDX78avWu1yOA87Z1v0MDs70diW9Y6IHk3uxfv16\nLr30kjpKAmDvvfdhwYIFxY7X3gfHRf8aHJzH/Pn3/Jno9/FgHzayF5VuPf/pBI/PAydExA8z84bJ\nKyPigcAJwKc7VRzVaZsVrdAx4VfAP01nJ8PDQx0saW6b3IuLLrqcN5z8ZRYtWVq8ltUrr+WTJwyx\n3377FT/28PCQ46KPDQ8PsXjxjpssk31oZy+6YzrB48PAkcBlEfFpqk+nXQUsobrg8yjgd8ApHazv\nBuDhETE/Mze0li0DrprOTkZG1jE62tHbi8w5g4PzGB4e2qQXIyPrWLRkKTvvukctdY2MrGPVqrXF\njtfeh5GRrZ05VK9qH3eb+9noN/ZhI3tRmehDp21z8MjM0Yj4S6pZjaOB9ne33EwVTE7MzE6+mn8L\nOAn4VEScCDyS6h0t03pXy+joGBs29O/gaTe5F3X/UNX1bzM6Olb7c1d9php3vk5U7MNG9qI7pnXn\n0tYpjzdFxD8BuwP3Bv4EXJmZ450uLjNHWtd5fBD4RetY78zMT3X6WJIkqftm9FktrdMev+twLZs7\n1hVUt2aXJElzXH9fsitJkooyeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKkYgwekiSpGIOHJEkq\nxuAhSZKKMXhIkqRiDB6SJKkYg4ckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKk\nYgwekiSpGIOHJEkqxuAhSZKKMXhIkqRiDB6SJKmY+XUXsDURsQD4APBc4E7g05n51nqrkiRJMzEX\nZjw+BDwZeArwPOAlEfGSekuSJEkz0ejgERGLgaOBF2fmf2Xmj4BTgAPqrUySJM1E00+1HALcmpkX\nTCzIzJNqrEeSJM1C04PH7sDVEfEC4J+ABcBngBMzc7zWyiRJ0rQ1PXgsBB4BvAQ4Crg/8AlgDXDa\ntu5kcLDRZ5SKmOjB5F7U2Zux0Q2sWJFFa5g3b4CFC7dnzZo7WLEiix1XzTI4OI/58+/5M9HvrxP2\nYSN7UenW82968NgALAKel5nXAUTEQ4BjmEbwGB4e6k51c9DkXtTZm7W33sgnv3UDi362upbj33zl\nReyy+361HFv1Gh4eYvHiHTdZJvvQzl50R9ODx43AHROhoyWBB09nJyMj6xgdHetoYXPN4OA8hoeH\nNunFyMi6GquCRUuWsvOue9Ry7NUr/1DLcVW/kZF1rFq1Ftj8z0a/sQ8b2YvKRB86renB4+fA9hHx\n8Mz8n9ayPYGrp7OT0dExNmzo38HTbnIv+vmHSv1rqtcEXycq9mEje9EdjT6BlZkrgHOBz0bEoyLi\nqcCbgY/WW5kkSZqJps94ADwfOB34KXA7cHpmfqTekiRJ0kw0Pnhk5mqqd7QcVW8lkiRpthp9qkWS\nJPUWg4ckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKkYgwekiSpGIOHJEkqxuAh\nSZKKMXhIkqRiDB6SJKkYg4ckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKkYgwe\nkiSpGIOHJEkqZk4Fj4g4NyI+XXcdkiRpZuZM8IiI5wCH112HJEmauTkRPCJiMXAS8Iu6a5EkSTM3\nv+4CttEpwFnAA+suRJIkzVzjZzwi4knA44ET6q5FkiTNTqNnPCJiO+BjwCsy886ImNF+Bgcbn6+6\nbqIHk3thb9SPBgfnMX/+PX8mSv0srF+/nksvvaTIsaay9977sGDBgk2Wl+5Dk9mLSreef6ODB/AO\n4OLM/MFsdjI8PNSZanrA5F7YG/Wj4eEhFi/ecZNlJVx00eW84eQvs2jJ0iLHa7d65bV88oQh9ttv\nv81u42vCRvaiO5oePJ4N7BIRq1uPtwOIiCMzc3hbdzIyso7R0bFu1DdnDA7OY3h4aJNejIysq7Eq\nqR4jI+tYtWotsPmfjW4ee9GSpey86x5dP9bmjj/x3NuV7kOT2YvKRB86renB41DgXm2PTwLGgTdN\nZyejo2Ns2NC/g6fd5F708w+V+tdUrwmlXifq/pnb2vP09XIje9EdjQ4emfmH9setmY/xzLyqppIk\nSdIs9PeVM5IkqahGz3hMlpkvqrsGSZI0c854SJKkYgwekiSpGIOHJEkqxuAhSZKKMXhIkqRiDB6S\nJKkYg4ckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJKmZOfVaLJM3W2OgGMq+4+/Hg4DyGh4cYGVlX\n5CPr248t9SODh6S+svbWGznz3BtYdOGaWo5/85UXscvu+9VybKkJDB6S+s6iJUvZedc9ajn26pV/\nqOW4UlN4jYckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKkYgwekiSpGIOHJEkq\npvF3Lo2IBwAfAp4I3A58EXhLZq6vtTBJkjRtjQ8ewFeAlcDBwBLgM8AG4M11FiVJkqav0cEjIgLY\nH9glM29pLXs7cDIGD0mS5pymX+NxE3D4ROhoGQB2qqkeSZI0C42e8cjM24DvTTyOiAHglcAPaitK\nkiTNWKODxxROBh4NPHY63zQ42OyJne/94Ies/POfu3qMgXkDDG1/L9bdcRfjY+N3L7/2mquAhV09\ntqTmGBycx/z5m74mTrxONv31sgR7UenW858zwSMi3ge8GnhWZv73dL53eHioO0V1yFlf+T7rd1pe\n4Eijmyy5+YoV7PSgEseW1ATDw0MsXrzjFterYi+6Y04Ej4g4HXgZ8PzM/Pp0v39kZB2jo2OdL6xD\n5s1fwPYL713Lsecv2KGW40qqx8jIOlatWrvJ8sHBeQwPDzX+9bIEe1GZ6EOnNT54RMTxwEuBZ2fm\n12ayj9HRMTZsaPDgGd/6JpLUCVt7PWz862VB9qI7Gh08ImIZcBzwbuBnEbHLxLrMvLm2wiRJ0ow0\n/cqZp1PVeBxwQ+vPja2/JUnSHNPoGY/MfB/wvrrrkCRJndH0GQ9JktRDDB6SJKkYg4ckSSrG4CFJ\nkooxeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKkYgwekiSpmEbfMl2S1DljoxvIvGLKdaU+Cn6v\nvfZhwYIFXdv/5qxfv57LLrtkm7btVi/qeu5NY/CQpD6x9tYbOfPcG1h04Zpajr965bWc9HpYvvwx\nxY992WWX8KZTv8qiJUuLHxvqfe5NY/CQpD6yaMlSdt51j7rLqEU/P/cm8RoPSZJUjMFDkiQVY/CQ\nJEnFGDwkSVIxBg9JklSMwUOSJBVj8JAkScUYPCRJUjEGD0mSVEzj71waEdsBHwWOAG4H3p+Zp9Zb\nlSRJmom5MONxCrAvcBjwCuD4iDii1ookSdKMNDp4RMQOwD8Ar87M32TmN4CTgFfWW5kkSZqJRgcP\n4C+oTgf9vG3ZBcAB9ZQjSZJmo+nB4/7ALZm5oW3ZzcD2EbGkppokSdIMNf3i0h2AOyctm3i83bbu\nZHCw4flqoN7Dr155bS3Hvf22m4DxWo7d78f3ufvc67B65bWsWLGoltfkFSuyttc6qJ774OD+zJ/f\n8N9Hbbr17zQwPl7fINyaiDgS+FBmPqBt2SOBy4AlmXlrbcVJkqRpa3r0uh64T0S017krsM7QIUnS\n3NP04PFr4C7gwLZljwcuqqccSZI0G40+1QIQER8DDgaOBh4EfBY4KjO/XmddkiRp+pp+cSnA66nu\nXPpD4DbgbYYOSZLmpsbPeEiSpN7R9Gs8JElSDzF4SJKkYgwekiSpGIOHJEkqxuAhSZKKmQtvp52R\niNiO6m24RwC3A+/PzFPrrar7IuKZwFepPpBhoPX3VzLzWRGxHPgYsA9wKXBMZv6ytmK7pPVvfzFw\nbGae31r2UOCTwOOAq4HXZeb3277nL4EPALtTfRrySzLzqrKVd9Zm+vBB4FXcc3y8KjM/2lr/XOAE\nqjsEf4+qDytrKL8jIuIBwIeAJ1K9DnwReEtmru+nMbGVPvTbmHgY8BGq+0OtBD6cmae01j2U/hkT\nW+pDV8dEL894nALsCxwGvAI4PiKOqLWiMvYEvkk1IHal+oTfF0fEDsC5wE+o+vJz4NyIGKqr0G5o\n/bI9h6oP7b4O3AA8Bvg88LWIeFDrex4MfA04E3gscEtr+zlrC31YBryZalxMjI9Pt75nf+BTwPFU\ndwteTHXDvrnsK8D2VC+uzwH+huoFE+Ab9M+Y2FIf+mZMRMQA1evgzcCjgZcDx0XEc1qb9MWY2IY+\ndHVM9OSMR+uX7D8AT83M3wC/iYiTgFdSzQb0smXApZn5p/aFEXE0cHtmvrm16LUR8VfA3wFnFa6x\nKyJiGXD2FMufRPU/lAMz8w7gvRHxZKq74b4TeAlwUWae1tr+RcBNEfGEiZmCuWRzfWhZBpyUmX+c\nYt2xwBcy819b+3kBcE1EPCQzr+lOtd0TEQHsD+ySmbe0lr0dODki/g3YDTig18fElvpA9culb8YE\nsAvwK+AVmbkW+H1EnAccEhE30ydjgi30Afh/dHlM9OqMx19Qhaqfty27ADignnKK2hP43RTLD6Dq\nQbv/oJpS7BWHAudRPaeBtuUHAL9svZhMuICNz/0A4O4XjsxcB/ySudubKfsQEYuABzL1+IDqfy/t\nfbgOuJZ7flbSXHITcPjEL9s2O1E9p34ZE1P1YQDYqd/GRGbelJnPbf2yJSIOpvr8rx/TR2NiM314\nAvCjEmOiJ2c8qKaFbsnMDW3Lbga2j4glc/n85DYI4GkR8VZgkOpc7vFUPbl00rY3A3uVLa97MvOM\nia+r/+Td7f5U06ftbqb67J9tWT+nbKEPy6jO1R4XEYdTndc9NTMnZrx6rQ+3UZ1/Bu6eXn4lVSjr\nmzGxhT78gD4bE+0i4mrgwcC3qWbCT6NPxkS7KfqwP10eE70647EDcOekZROPtytcSzERsRQYAtZR\nnUJ5A/B8qinVzfWkZ/vRZmvPvV9680hgDLgcOJzqPO0nIuIZrfW93oeTgeXAW+nvMXEy1Xn94+jv\nMXEE1bUuj6a6YLRfx8REH5ZTha+gy2OiV2c87mDTJkw8vr1wLcVk5rWtGZ1bW4t+GxGDVBdJ/Yip\ne9Kz/WhzB3DvScvan/vmxsuqLtdVVGaeFRHfbBsfl0bEI4BjqC6q21wf5vwYiYj3Aa8GnpWZl0dE\nX46JyX0ALu/XMTHxjr6IeD3wr1QXjS6etFnPj4m2PryO6nfFMNDVMdGrMx7XA/eJiPbntyuwrq2Z\nPWmK5/ffVFez30TVg3a7AjeWqKtm17Pl57619T1jM+Pjga2ve7IPEXE68Drg+W2fbN13Y2Izfeir\nMRER92v7n/uEy4EFVM+pL8bEVvqwqNtjoleDx6+Bu7jnxS6PBy6qp5wyIuJ/RcQtEbF92+LlVG/7\n+inVW+naHQRcWKq+Gl0I7Nt6i+mEQ9j43C9sPQbuflfUcnqsNxHxzxHx/UmLlwNXtL6e3IcHU523\nnbN9iIjjgZcCz87ML7Wt6qsxsbk+9OGY2A34akS0/+J8LPBHqgtJH9MnY2JzffgT8Jpuj4mB8fHx\nmRTdeBHxMapftEdTNeWzwFHtSb/XRMRCqtR6PtXbvx5GdTOcDwCfAP6H6m2Wn6B63/aRwMNbV2f3\nlIgYAw7LzPNbM1+/obq49gTg6cBbgL0y87qIeAhV3/6Z6gKr44E9MnPfeqrvnEl9eCzVO5neQnX/\ngacC72+z/0aQAAAFs0lEQVSt/0VEHEh1Su5YqhuPnQbclpl/W0/1s9N6W/FvgXdT3Uyw3Z/okzGx\nlT48mP4aE/Oo3u34Z+D1VL+AP8XG3vwWuITeHxNb6sOFdHlM9OqMB1TN/C/gh8DpwNt6OXQAZOYa\nqkFyX6rZnU8CZ2Tm+zNzNfDXVG+ZupjqyuXDezF0tNydqDNzDHgG1XTgxcDzgGe23gZG673nR1CF\n1F8AOwNz8oV1Cu19uJgqbL6Q6sX1lcBzM/MXrfUXAi+jekG9gOpq9qNLF9xBT6d6jTuO6ir8G6im\ng29ojYln0h9jYkt96Ksx0fZasBb4GdV/wj6YmR9urXs6fTAmttKHro+Jnp3xkCRJzdPLMx6SJKlh\nDB6SJKkYg4ckSSrG4CFJkooxeEiSpGIMHpIkqRiDhyRJKsbgIUmSijF4SJKkYubXXYCk7oiIH1Pd\nIn8q48B9M/PP5SraKCKuAj6Tme+s4/htdewJPDQzv9N6PEb1mU5n1VmX1Muc8ZB61zjwBWAXqs+f\naP9z/7pCR8N8m+pTOSUV4oyH1NvWZeaf6i6iwQbqLkDqNwYPqY9FxP5UH4H9xsw8rbXs3cAxwKMz\n85rWx6q/FziY6jXj+8AbMvPa1vY/ovoo7fsC/4dqJvVbwMsyc+0sansR8EbgocBVwMeB0zNzvPUR\n5VdRfYrmm4C/AG4C3p2Zn2zbx2uBV1PN+lzQ+nN0Zu7WOt2zFHhHRByWmU9qfdsjI+L7wCFUn7z5\n4cx870yfh6R78lSL1MdaH3X9buCEiNgtIg6h+kX+slboWEr1sdnrgEOBp1Cdqjk/Iha27eq1VB+1\n/ljg+VQfOf+6mdYVES8FTqL66O09qT7S/c3AeyZteipwArCM6rTJR1uhhIg4FngX8M9UweTHrf1N\nfCT3fsD1wCnc8+PNjwU+09rnx4B3R8RhM30uku7JGQ+pt/19RPzdpGXjwNcy8/+2Hr8TeBrwKWA3\n4LOZ+cXWumOB1cALMvMugIg4kmq24e+BM1rbXZ6Zb2t9/fvWjMHBs6j7OOCEzPxS6/HVEbETVbB4\ne9t278/Mc1t1Hdeq90DgGuAfgdMy819a274nIh4LLAfIzFsiYhRYk5m3te3zI5l5duvrEyPiH6kC\n1Y9n8XwktRg8pN72DaoZjMnXMqyZ+CIzRyPiBcAlVKcrXtW23d7AxROho7X9zRGRwD5t210xaf+3\nUp3GmLaIuA/wIKqgcGLbqnnAAqpwdMfk42bmbREBsCAi7g08hOoUULvzaQWPLVgx6fGtwNB0noOk\nzTN4SL1tdWZetQ3bPYrqF/v9W1//Z2v5ABtPTbSbB9zV9vjOKbaZ6YWbE6eAXwucN8X6a4EHbuW4\nGybtazpGN7NPSR3gNR5Sn4uI+1Ndy/Au4Gzg8xEx8T/83wL7R8S92rbfBdgDuKwb9WTmH4E/Ag/L\nzCsn/lBdk3Ei2xACMnOE6nTLgZNWHTTp8VShSlIXOeMh9bahVlCYyqrMXA98GriO6pf6IuBy4APA\ny6kCycuBz7VOewwBJ1MFgy/MsrY9IuKpk5aty8zzqS4sfVdE/AH4LtXFoR+lujblrtYpla15L3BK\nRFxB9c6dv6V61801bdusadVxv1bgkdRlBg+ptz2r9afdxOmTv4uIXYEnAwdk5gZgVUS8EvhiRHwj\nM78bEYdSBYGfU53a+Hfg+a1Zhdl4XutPu2uA3TPz1Ii4neqtsO+nuvbkDOAdbdtONVtx97LM/HhE\nLKaaybkP1cWhn6F6m+yED1EFqb2prv3Y4j4lzd7A+Lg/U5J6T2s25bLMvK5t2cepgs1T6qtM6m/O\neEjqVS8ElkXEMVT3GDmM6h4jx9RZlNTvDB6SetWxVDcY+xqwM/B74DWZ+blaq5L6nKdaJElSMb6d\nVpIkFWPwkCRJxRg8JElSMQYPSZJUjMFDkiQVY/CQJEnFGDwkSVIxBg9JklTM/wezBsoIeuOwkAAA\nAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10d568110>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "exon_lengths = [coord[1] - coord[0] for exon, coord in exon_dict.items() if exon != '1']\n",
    "print 'Exon 1 Coordinates: {}'.format(exon_dict['1'])\n",
    "plt.figure()\n",
    "plt.hist(exon_lengths, bins=12)\n",
    "plt.xlabel('Exon Length'), plt.ylabel('Count')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Playing around with pysam and samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "bamfile = pysam.AlignmentFile('../bams/FGPC-0000-444F.bam', 'rb')\n",
    "print bamfile.references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SQ', 'RG', 'PG', 'HD']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'ID': 'FGPC-0000-444F-del_1-M1-1_H5FGFAFXX-L004',\n",
       "  'LB': 'FGPC-0000-444F_1-M1',\n",
       "  'PL': 'ILLUMINA',\n",
       "  'PU': 'H5FGFAFXX-L004',\n",
       "  'SM': 'FGPC-0000-444F'},\n",
       " {'ID': 'FGPC-0000-444F-del_1-M1-1_H5FGFAFXX-L003',\n",
       "  'LB': 'FGPC-0000-444F_1-M1',\n",
       "  'PL': 'ILLUMINA',\n",
       "  'PU': 'H5FGFAFXX-L003',\n",
       "  'SM': 'FGPC-0000-444F'},\n",
       " {'ID': 'FGPC-0000-444F-del_1-M1-1_H5FGFAFXX-L002',\n",
       "  'LB': 'FGPC-0000-444F_1-M1',\n",
       "  'PL': 'ILLUMINA',\n",
       "  'PU': 'H5FGFAFXX-L002',\n",
       "  'SM': 'FGPC-0000-444F'},\n",
       " {'ID': 'FGPC-0000-444F-del_1-M1-1_H5FGFAFXX-L001',\n",
       "  'LB': 'FGPC-0000-444F_1-M1',\n",
       "  'PL': 'ILLUMINA',\n",
       "  'PU': 'H5FGFAFXX-L001',\n",
       "  'SM': 'FGPC-0000-444F'}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# header breakdown\n",
    "print bamfile.header.keys()\n",
    "# take template header from existing bamfile\n",
    "template_header = bamfile.header\n",
    "for RG in template_header['RG']:\n",
    "    subject = '{}-del'.format(RG['ID'].split('_')[0])\n",
    "    RG['ID'] = '_'.join([subject] + RG['ID'].split('_')[1:])\n",
    "template_header['RG']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NS500214:178:H5FGFAFXX:1:11101:1977:9301\t147\t22\t31854685\t60\t151M\t22\t31854633\t151\tAAGCTTTGCCTCTTCTATTACAGTATGGCCAGTATTTCCTTACAAGTTATTTCACTGATTATAAATAGTCCACGTCAATGGCAAATGTACAACAGGGGAAGCATAACCCATTATGAGGTAATGGATATTGCTAGAGGTTGCTTCATTACCT\tarray('B', [12, 23, 18, 28, 27, 28, 25, 27, 29, 12, 26, 29, 29, 10, 33, 24, 29, 29, 27, 28, 30, 31, 26, 27, 29, 25, 9, 31, 32, 13, 32, 31, 27, 29, 24, 29, 28, 32, 33, 24, 27, 26, 34, 30, 32, 31, 29, 27, 29, 29, 29, 28, 34, 28, 33, 10, 29, 30, 30, 23, 30, 23, 31, 31, 30, 28, 33, 25, 29, 33, 35, 29, 28, 33, 29, 29, 31, 30, 31, 31, 32, 35, 31, 31, 30, 30, 33, 28, 29, 35, 31, 29, 35, 33, 31, 32, 25, 35, 32, 34, 33, 36, 31, 23, 32, 30, 34, 34, 36, 31, 31, 23, 31, 31, 35, 34, 32, 33, 29, 32, 31, 31, 32, 35, 31, 29, 31, 24, 31, 33, 35, 29, 34, 35, 34, 32, 33, 31, 31, 33, 35, 31, 30, 36, 31, 31, 28, 30, 29, 31, 29])\t[('BD', 'KMNLCKONPONMMNNLLKMLMMMKNMMLMMMMKKBLMNLJMLLKMKJKKBLNMMNLNKIJKJDKJKMMLLLLKMMKJMLLLKCJMJLLKKJKLLIILKJLLMJIJKHLMJIJMLKLKMJJNMMNKKKJNNMLLLNMLKPPNNQQNMPLNKK'), ('MD', '0C12G137'), ('BI', 'OQPODKOQQQQQQQQPPOQPRQPPRRQRQRQPPOGPRPONPOOOQMOOOGOQPQPPQONOOOHNOOPPQPPQOPQONQQPQOHNQNOPOONOQQMNPOOQPQOOMOLPQONNQPMQNOONQQPQOOOMPPPOOMQPNNQRQQSSQQSPPNN'), ('NM', 2), ('AS', 145), ('XS', 0), ('RG', 'FGPC-0000-444F_1-M1-1_H5FGFAFXX-L001')]\n"
     ]
    }
   ],
   "source": [
    "# does not output in same format as actual samfile -- see documentation for breakdown\n",
    "iterable = bamfile.fetch('X', *EXON49)\n",
    "example_read = iterable.next()\n",
    "print example_read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NS500214:178:H5FGFAFXX:1:11101:1977:9301\t99\t22\t31854633\t60\t151M\t22\t31854685\t151\tGTTAATAGCCTTAAGATCACAATATATAAATAGGATATGCTGATCAAAAAGACAGCTTTGCCTCTGCTATTACAGTATGGCCAGTATTTCCTTACAAGTTATTTCACTGATTATAAATAGTCCACGTCAATGGCAAATGTACAACAGGGGA\tarray('B', [27, 28, 28, 26, 27, 31, 29, 35, 32, 32, 34, 32, 28, 30, 35, 30, 31, 35, 31, 33, 31, 31, 31, 29, 31, 29, 31, 24, 31, 31, 31, 23, 35, 34, 30, 31, 29, 31, 36, 25, 34, 36, 30, 31, 35, 31, 31, 31, 30, 31, 35, 29, 32, 31, 34, 32, 34, 32, 32, 36, 33, 32, 34, 35, 31, 36, 33, 33, 28, 30, 32, 28, 33, 31, 34, 29, 28, 30, 35, 13, 20, 9, 30, 34, 29, 28, 30, 31, 31, 34, 9, 31, 31, 28, 32, 30, 30, 33, 28, 31, 28, 30, 31, 31, 34, 30, 31, 32, 25, 25, 30, 31, 18, 26, 25, 30, 30, 23, 28, 29, 29, 30, 9, 26, 17, 24, 28, 31, 29, 11, 23, 26, 19, 17, 26, 26, 29, 29, 34, 28, 27, 23, 26, 26, 18, 25, 28, 14, 24, 23, 21])\t[('BD', 'KKKJMNNPQOOMLLMNNOMLKJKKKJJJIAJJLMLLJJMLLLMLMLIAAAKLMJMMLJCKLLLKKLLLKJJILJMMLJMLKLLMMLJJCKLLKJMJJLNKJKKDLMLMMNMKJKKJBKKMNNMMLLMNMJKNMMOKCLOMNOMLMLONIIM'), ('MD', '151'), ('BI', 'NNNOQRRSRRSQPOPPQQPONMOONONONGNNOOPPNNPPPPPPPOLFFFNNOMOOPNGNPPQMNPPPNNNOONPQPOQPQPQPQPONHOPQOOONMOQNOONHOPOPQQQNOOONGOPPRRQRPPRRQNPRQRQNHQSQQPOMLLNPLLQ'), ('NM', 0), ('AS', 151), ('XS', 0), ('RG', 'FGPC-0000-444F_1-M1-1_H5FGFAFXX-L001')]\n",
      "True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# keep track of qnames, match between read pairs\n",
    "print example_read_mate\n",
    "print example_read_mate.is_read1\n",
    "example_read_mate.qname == example_read.qname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('BD',\n",
       "  'KMNLCKONPONMMNNLLKMLMMMKNMMLMMMMKKBLMNLJMLLKMKJKKBLNMMNLNKIJKJDKJKMMLLLLKMMKJMLLLKCJMJLLKKJKLLIILKJLLMJIJKHLMJIJMLKLKMJJNMMNKKKJNNMLLLNMLKPPNNQQNMPLNKK'),\n",
       " ('MD', '0C12G137'),\n",
       " ('BI',\n",
       "  'OQPODKOQQQQQQQQPPOQPRQPPRRQRQRQPPOGPRPONPOOOQMOOOGOQPQPPQONOOOHNOOPPQPPQOPQONQQPQOHNQNOPOONOQQMNPOOQPQOOMOLPQONNQPMQNOONQQPQOOOMPPPOOMQPNNQRQQSSQQSPPNN'),\n",
       " ('NM', 2),\n",
       " ('AS', 145),\n",
       " ('XS', 0),\n",
       " ('RG', 'FGPC-0000-444F_1-M1-1_H5FGFAFXX-L001')]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_read.get_tags()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "coverage = bamfile.count_coverage('X', *EXON49)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "104\n",
      "array('L', [0L, 0L, 1L, 260L, 1L, 0L, 0L, 0L, 0L, 0L, 0L, 281L, 0L, 0L, 0L, 0L, 1L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 2L, 0L, 0L, 0L, 0L, 299L, 0L, 287L, 284L, 285L, 1L, 0L, 1L, 0L, 0L, 2L, 0L, 0L, 0L, 1L, 0L, 319L, 0L, 308L, 1L, 315L, 323L, 307L, 305L, 0L, 1L, 0L, 0L, 0L, 0L, 0L, 2L, 301L, 1L, 303L, 0L, 0L, 1L, 1L, 0L, 1L, 0L, 0L, 0L, 0L, 0L, 339L, 0L, 0L, 0L, 0L, 0L, 326L, 303L, 0L, 1L, 0L, 0L, 0L, 289L, 0L, 0L, 0L, 0L, 307L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L])\n"
     ]
    }
   ],
   "source": [
    "print len(coverage[0])\n",
    "print coverage[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'X'"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# getting reference name\n",
    "bamfile.getrname(example_read.reference_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.68 s, sys: 107 ms, total: 6.79 s\n",
      "Wall time: 6.84 s\n"
     ]
    }
   ],
   "source": [
    "# next two cells testing pysam's samtools integration (specifically samtools view and merge)\n",
    "%%time\n",
    "infile = 'FGPC-8V7M-82J3.bam'\n",
    "outfile = 'FGPC-8V7M-82J3_ch22.bam'\n",
    "pysam.view('-b', '-o' + outfile, infile, '21', '22', catch_stdout=False)\n",
    "pysam.index(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 14 s, sys: 192 ms, total: 14.2 s\n",
      "Wall time: 14.3 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "infile = 'FGPC-8V7M-82J3.bam'\n",
    "outfile = 'FGPC-8V7M-82J3_ch20.bam'\n",
    "pysam.view('-b', '-o' + outfile, infile, '20', catch_stdout=False)\n",
    "# pysam.index(outfile)\n",
    "\n",
    "pysam.merge('combined.bam', outfile, 'FGPC-8V7M-82J3_ch22.bam', catch_stdout=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recording Alignment Statistics and Insert Lengths\n",
    "\n",
    "Low mapping quality could indicate both external homology and self-similarity in the DMD gene, which could complicate techniques based on read coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def alignment_stats(bamfile_name, reference, region, minmapQ=30):\n",
    "    '''Prints alignment stats across reads in particular region in BAM. \n",
    "       reference -- ref or chrm number\n",
    "       region -- list or tuple with region bounds\n",
    "    '''\n",
    "    with pysam.AlignmentFile(bamfile_name, 'rb') as bamfile:\n",
    "        stats = Counter()\n",
    "        print '{} \\nChromosome {} Region {} Stats:'.format(bamfile_name, reference, region)\n",
    "        for read in bamfile.fetch(reference, *region):\n",
    "            stats['total'] += 1\n",
    "            stats['qcfail'] += int(read.is_qcfail)\n",
    "\n",
    "            stats['paired'] += int(read.is_paired)\n",
    "            stats['read1'] += int(read.is_read1)\n",
    "            stats['read2'] += int(read.is_read2)\n",
    "            \n",
    "            if read.is_unmapped:\n",
    "                stats['unmapped'] += 1\n",
    "                continue\n",
    "\n",
    "            stats['mapq <= 30'] += int(read.mapping_quality <= minmapQ)\n",
    "\n",
    "            stats['mapped'] += 1\n",
    "            if read.is_proper_pair:\n",
    "                stats['proper pair'] += 1\n",
    "                if abs(read.template_length) > 850:\n",
    "                    stats['long_insert'] += 1\n",
    "                    print '\\tlong insert/error: {}'.format(read.template_length)\n",
    "                else:\n",
    "                    stats['insert_lengths'] += (abs(read.template_length) if read.is_read1 else 0)\n",
    "\n",
    "        stats['avg insert length'] = stats['insert_lengths'] / float(stats['read1'])\n",
    "        output_order = ('total', 'mapped', 'unmapped', 'paired', 'read1', 'read2', \n",
    "                        'proper pair', 'qcfail', 'mapq <= 30', 'long_insert', 'avg insert length')\n",
    "\n",
    "        for key in output_order:\n",
    "            if key == 'avg insert length':\n",
    "                print '\\t{}: {}'.format(key, stats[key])\n",
    "            else:\n",
    "                print '\\t{}: {}, Prop: {}'.format(key, stats[key], stats[key] / float(stats['total']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bams/FGPC-0000-444F.bam \n",
      "Chromosome X Region [31854835, 31854939] Stats:\n",
      "\ttotal: 680, Prop: 1.0\n",
      "\tmapped: 680, Prop: 1.0\n",
      "\tunmapped: 0, Prop: 0.0\n",
      "\tpaired: 680, Prop: 1.0\n",
      "\tread1: 342, Prop: 0.502941176471\n",
      "\tread2: 338, Prop: 0.497058823529\n",
      "\tproper pair: 680, Prop: 1.0\n",
      "\tqcfail: 0, Prop: 0.0\n",
      "\tmapq <= 30: 0, Prop: 0.0\n",
      "\tlong_insert: 0, Prop: 0.0\n",
      "\tavg insert length: 226.011695906\n"
     ]
    }
   ],
   "source": [
    "alignment_stats('bams/FGPC-0000-444F.bam', 'X', EXON49)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bams/FGPC-0000-444F.bam \n",
      "Chromosome X Region [31115794, 33357558] Stats:\n",
      "\tlong insert/error: 87733677\n",
      "\tlong insert/error: 36406070\n",
      "\tlong insert/error: 67357711\n",
      "\ttotal: 88693, Prop: 1.0\n",
      "\tmapped: 88577, Prop: 0.998692117755\n",
      "\tunmapped: 116, Prop: 0.0013078822455\n",
      "\tpaired: 88693, Prop: 1.0\n",
      "\tread1: 44343, Prop: 0.499960538036\n",
      "\tread2: 44350, Prop: 0.500039461964\n",
      "\tproper pair: 88301, Prop: 0.995580259998\n",
      "\tqcfail: 0, Prop: 0.0\n",
      "\tmapq <= 30: 180, Prop: 0.00202947244991\n",
      "\tlong_insert: 3, Prop: 3.38245408319e-05\n",
      "\tavg insert length: 246.341519518\n",
      "bams/FGPC-8V7M-82J3.bam \n",
      "Chromosome X Region [31115794, 33357558] Stats:\n",
      "\tlong insert/error: 8528942\n",
      "\tlong insert/error: -3886263\n",
      "\ttotal: 19239, Prop: 1.0\n",
      "\tmapped: 19235, Prop: 0.999792088986\n",
      "\tunmapped: 4, Prop: 0.000207911014086\n",
      "\tpaired: 19239, Prop: 1.0\n",
      "\tread1: 9621, Prop: 0.50007796663\n",
      "\tread2: 9618, Prop: 0.49992203337\n",
      "\tproper pair: 19209, Prop: 0.998440667394\n",
      "\tqcfail: 0, Prop: 0.0\n",
      "\tmapq <= 30: 29, Prop: 0.00150735485212\n",
      "\tlong_insert: 2, Prop: 0.000103955507043\n",
      "\tavg insert length: 232.779232928\n",
      "bams/FPWB-0000-840T.bam \n",
      "Chromosome X Region [31115794, 33357558] Stats:\n",
      "\tlong insert/error: 45328884\n",
      "\tlong insert/error: 42059688\n",
      "\tlong insert/error: 90499310\n",
      "\tlong insert/error: 109798660\n",
      "\tlong insert/error: 120075025\n",
      "\tlong insert/error: 34330200\n",
      "\ttotal: 116333, Prop: 1.0\n",
      "\tmapped: 116191, Prop: 0.99877936613\n",
      "\tunmapped: 142, Prop: 0.00122063387001\n",
      "\tpaired: 116333, Prop: 1.0\n",
      "\tread1: 58181, Prop: 0.500124642191\n",
      "\tread2: 58152, Prop: 0.499875357809\n",
      "\tproper pair: 115867, Prop: 0.995994257863\n",
      "\tqcfail: 0, Prop: 0.0\n",
      "\tmapq <= 30: 293, Prop: 0.00251863185854\n",
      "\tlong_insert: 6, Prop: 5.15760790146e-05\n",
      "\tavg insert length: 302.194410546\n",
      "bams/MGPC-6KHG-N0LS.bam \n",
      "Chromosome X Region [31115794, 33357558] Stats:\n",
      "\tlong insert/error: 79779343\n",
      "\tlong insert/error: -1226\n",
      "\tlong insert/error: -1476\n",
      "\ttotal: 18874, Prop: 1.0\n",
      "\tmapped: 18867, Prop: 0.999629119424\n",
      "\tunmapped: 7, Prop: 0.000370880576454\n",
      "\tpaired: 18874, Prop: 1.0\n",
      "\tread1: 9437, Prop: 0.5\n",
      "\tread2: 9437, Prop: 0.5\n",
      "\tproper pair: 18822, Prop: 0.997244887146\n",
      "\tqcfail: 0, Prop: 0.0\n",
      "\tmapq <= 30: 82, Prop: 0.00434460103847\n",
      "\tlong_insert: 3, Prop: 0.00015894881848\n",
      "\tavg insert length: 278.86372788\n"
     ]
    }
   ],
   "source": [
    "for bam in bamfiles:\n",
    "    alignment_stats('bams/{}'.format(bam), 'X', DMD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusions\n",
    "Mapping quality seems to be fairly high across DMD gene (only 0.1-0.4% of reads have mapQ less than 30) for 4 samples tested here. \n",
    "\n",
    "Repeat with more samples -- could also check specific regions, add check to see if reads with low mapQ are also short or low quality reads. \n",
    "\n",
    "In addition, dot plots of several different regions of the gene (region compared to itself) indicate few patterns of self-similarity within the gene.\n",
    "\n",
    "Pixel gradient is based on proportion of matching bases with window size 25. No gradation below 20% homology. See plot titles for region coordinates.\n",
    "\n",
    "<img src=\"images/region1_dot.png\" width=\"500\">\n",
    "<img src=\"images/region2_dot.png\" width=\"500\">\n",
    "<img src=\"images/region3_dot.png\" width=\"500\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing basic BAMs with simulated deletions and duplications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def simulate_exon_CNV(outfile, infile_bam, chrom_ref, mut_region, CNV_type='deletion', region_buffer=0, mult_factor=0.5, \n",
    "                      output_full=False, before_wanted=None, after_wanted=None):\n",
    "    ''' Simulates deletion in specified region by reducing the number of aligned reads in region \n",
    "        by specified depletion_factor. Assumes bam file is sorted.\n",
    "        chrom_ref -- '1', 'X', etc\n",
    "        mut_region -- two element list with region limits for region to be deleted/duplicated, eg. [34000000, 34560000]\n",
    "        CNV_type -- 'deletion' or 'duplication'\n",
    "        region_buffer -- number of bp by which to extend region on either end (reads often extend past exons)\n",
    "        mult_factor -- float indicating factor to multiply by read count (should be > 1 for duplications)\n",
    "        output_full -- option to generate complete bam file (all chromosomes) vs bam file with just chrom_ref\n",
    "    '''\n",
    "    if mut_region[0] > mut_region[1]:\n",
    "        raise ValueError('Region boundaries invalid')\n",
    "    mut_region = [mut_region[0] - region_buffer, mut_region[1] + region_buffer]\n",
    "        \n",
    "    template_bam = pysam.AlignmentFile(infile_bam, 'rb')\n",
    "    template_header = template_bam.header\n",
    "    for RG in template_header['RG']:\n",
    "        RG['SM'] = '{}_{}'.format(RG['SM'], ('del' if CNV_type == 'deletion' else 'dup'))\n",
    "    \n",
    "    with pysam.AlignmentFile(outfile, 'wb', header=template_header) as outf:\n",
    "        keep_reads = []\n",
    "        total_reads = []\n",
    "        started_mut = False\n",
    "        finished_mut = False\n",
    "        for read in template_bam.fetch(chrom_ref):\n",
    "            if (\n",
    "                (read.reference_end >= mut_region[0] or started_mut) and \n",
    "                read.reference_start <= mut_region[1]\n",
    "               ):\n",
    "                if not started_mut:\n",
    "                    started_mut = True\n",
    "                if read.is_paired and read.is_read1:\n",
    "                    if random.random() < (mult_factor if CNV_type == 'deletion' else mult_factor - 1.0):\n",
    "                        keep_reads.append(read.qname)\n",
    "                total_reads.append(read)\n",
    "                    \n",
    "            elif read.reference_start > mut_region[1]:\n",
    "                if finished_mut:\n",
    "                    outf.write(read)\n",
    "                else:\n",
    "                    print 'Writing to outfile from mutation region'\n",
    "                    for saved_read in total_reads:\n",
    "                        if CNV_type == 'duplication':\n",
    "                            outf.write(saved_read)\n",
    "                        if saved_read.qname in keep_reads:\n",
    "                            outf.write(saved_read)\n",
    "                    finished_mut = True\n",
    "                    outf.write(read)\n",
    "            else:\n",
    "                outf.write(read)\n",
    "        total_mate_names = [read.qname for read in total_reads if read.is_read2]\n",
    "        distant_mates = [name for name in keep_reads if name not in total_mate_names]\n",
    "        print 'Finished generating reads in bamfile for desired chrom_ref'\n",
    "        print 'Number of distant mates: {}'.format(len(distant_mates))\n",
    "        print 'Total number of mates to keep: {}'.format(len(keep_reads))\n",
    "    \n",
    "    # generates bam file for all chroms before desired and separate bam file for all chroms after; \n",
    "    # then merges three bam files -- might require clean up of extra files after running\n",
    "    if output_full:\n",
    "        ref_list = list(template_bam.references)\n",
    "        index = ref_list.index(chrom_ref)\n",
    "        \n",
    "        if not before_wanted:   \n",
    "            before = ref_list[:index]\n",
    "            before_wanted = '{}_before_{}.bam'.format(infile_bam.split('.')[0], chrom_ref)\n",
    "            pysam.view('-b', '-o' + before_wanted, infile_bam, *before, catch_stdout=False)\n",
    "            print 'Finished generating bamfile for refs before chrom_ref'\n",
    "            \n",
    "        if not after_wanted:\n",
    "            after = ref_list[(index + 1):]\n",
    "            after_wanted = '{}_after_{}.bam'.format(infile_bam.split('.')[0], chrom_ref)\n",
    "            pysam.view('-b', '-o' + after_wanted, infile_bam, *after, catch_stdout=False)\n",
    "            print 'Finished generating bamfile for refs after chrom_ref'\n",
    "\n",
    "        outfile_full = '{}_full.bam'.format(outfile.split('.')[0])\n",
    "        pysam.merge(outfile_full, before_wanted, outfile, after_wanted, catch_stdout=False)\n",
    "        print 'Finished merging all bamfiles into final output'\n",
    "        \n",
    "        template_bam.close()\n",
    "        pysam.index(outfile_full)\n",
    "    else:\n",
    "        pysam.index(outfile)\n",
    "        \n",
    "    print 'Finished creating index file for final output bamfile'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 1min, sys: 1.29 s, total: 1min 1s\n",
      "Wall time: 1min 2s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# assuming bams live in bams directory\n",
    "outfile = 'bams/FGPC-0000-444F_49_del.bam'\n",
    "simulate_exon_del(outfile, 'bams/FGPC-0000-444F.bam', 'X', EXON49, depletion_factor=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[31838092, 31838200]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_dict['30']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to outfile from mutation region\n",
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Number of distant mates: 32\n",
      "Total number of mates to keep: 128\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 58.9 s, sys: 1.16 s, total: 1min\n",
      "Wall time: 1min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# testing deletion\n",
    "# assuming bams live in bams directory -- this time with ensembl exon def\n",
    "outfile = 'bams/FGPC-0000-444F_30_delPEM.bam'\n",
    "simulate_exon_CNV(outfile, 'bams/FGPC-0000-444F.bam', 'X', exon_dict['30'], mult_factor=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to outfile from mutation region\n",
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Number of distant mates: 30\n",
      "Total number of mates to keep: 105\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 54.6 s, sys: 921 ms, total: 55.6 s\n",
      "Wall time: 55.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# testing duplication\n",
    "# assuming bams live in bams directory -- this time with ensembl exon def\n",
    "outfile = '../bams/FGPC-0000-444F_30_dupPEM.bam'\n",
    "simulate_exon_CNV(outfile, '../bams/FGPC-0000-444F.bam', 'X', exon_dict['30'], CNV_type='duplication', mult_factor=1.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to outfile from mutation region\n",
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Number of distant mates: 71\n",
      "Total number of mates to keep: 983\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 1min 13s, sys: 1.69 s, total: 1min 14s\n",
      "Wall time: 1min 15s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# another deletion\n",
    "# assuming bams live in bams directory -- this time with ensembl exon def\n",
    "outfile = 'bams/FPWB-0000-840T_45_delPEM.bam'\n",
    "simulate_exon_CNV(outfile, 'bams/FPWB-0000-840T.bam', 'X', exon_dict['45'], CNV_type='deletion', region_buffer=500,\n",
    "                  mult_factor=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to outfile from mutation region\n",
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Number of distant mates: 145\n",
      "Total number of mates to keep: 258\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 2min 11s, sys: 3.76 s, total: 2min 14s\n",
      "Wall time: 2min 18s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# another duplication\n",
    "# assuming bams live in bams directory -- this time with ensembl exon def\n",
    "outfile = 'bams/FPWB-0000-840T_19_dupPEM.bam'\n",
    "simulate_exon_CNV(outfile, 'bams/FPWB-0000-840T.bam', 'X', exon_dict['19'], CNV_type='duplication', mult_factor=1.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Finished generating bamfile for refs before chrom_ref\n",
      "Finished generating bamfile for refs after chrom_ref\n",
      "Finished merging all bamfiles into final output\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 7min 10s, sys: 7.01 s, total: 7min 17s\n",
      "Wall time: 7min 20s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# this is a relatively small bamfile (only about 1.5 GB) -- outputting full bamfile is quite slow in general\n",
    "infile = 'bams/FGPC-8V7M-82J3.bam'\n",
    "outfile = 'bams/FGPC-8V7M-82J3_48-50_del.bam'\n",
    "simulate_exon_del(outfile, infile, 'X', EX48_50_PART, depletion_factor=3, output_full=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Finished merging all bamfiles into final output\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 4min 3s, sys: 4.14 s, total: 4min 7s\n",
      "Wall time: 4min 9s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "infile = 'bams/FGPC-8V7M-82J3.bam'\n",
    "outfile = 'bams/FGPC-8V7M-82J3_48-50_del.bam'\n",
    "before_wanted = 'bams/FGPC-8V7M-82J3_before_X.bam'\n",
    "after_wanted = 'bams/FGPC-8V7M-82J3_after_X.bam'\n",
    "simulate_exon_del(outfile, infile, 'X', EX48_50_PART, depletion_factor=2, output_full=True, \n",
    "                  before_wanted=before_wanted, after_wanted=after_wanted)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Generating deletion of exons 48-50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to outfile from mutation region\n",
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Number of distant mates: 76\n",
      "Total number of mates to keep: 923\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 1min 5s, sys: 1.45 s, total: 1min 6s\n",
      "Wall time: 1min 7s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "ex48_50 = [exon_dict['48'][0], exon_dict['50'][1]]\n",
    "outfile = '../bams/FPWB-0000-822R_4850_delPEM.bam'\n",
    "simulate_exon_CNV(outfile, '../bams/FPWB-0000-822R.bam', 'X', ex48_50, CNV_type='deletion', mult_factor=0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating deletion of exons 16-19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to outfile from mutation region\n",
      "Finished generating reads in bamfile for desired chrom_ref\n",
      "Number of distant mates: 63\n",
      "Total number of mates to keep: 892\n",
      "Finished creating index file for final output bamfile\n",
      "CPU times: user 32.2 s, sys: 629 ms, total: 32.9 s\n",
      "Wall time: 33.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# note slightly higher mult_factor here\n",
    "ex16_19 = [exon_dict['16'][0], exon_dict['19'][1]]\n",
    "outfile = '../bams/FGPC-4ZPF-J0F1_1619_delPEM.bam'\n",
    "simulate_exon_CNV(outfile, '../bams/FGPC-4ZPF-J0F1.bam', 'X', ex16_19, CNV_type='deletion', mult_factor=0.65)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
